{: #intro}
High-throughput sequencing technologies have significantly improved the generation of 
large data sets of omics information (genomes, transcriptomes, proteomics and microbiomes). 
Administration and analysis of big data in the omics-era to understand 
complex biological systems present a new challenge for plant pathologists. 
Analysis of big and complex data sets requires use of computational tools 
that interact with databases and computer servers. 
The command line interface allows users to interact with computers both locally 
and in remote server to conduct computationally intensive analyses. 
In this workshop, we will explore different usages of command line interface for bioinformatic analyses. 
We will introduce the Linux shell, cluster computing system, and shell scripting. 
We will use unix commands and command line programs to run BLAST locally, parse sequence 
and develop phylogenetic tree. Different genome data analyses using 
command line interface will be discussed.