















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-10-20 23:33:00 -0400">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="https://bioinfoaps.github.io/">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    <link rel="license" href="#license-info" />

    

    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Basic	bioinformatics and command-line	tools	for phytopathologists"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  Phylogenetic tree &ndash; Basic	bioinformatics and command-line	tools	for phytopathologists
  </title>  

  </head>
  <body>

    <div class="container">
      
















  
  















<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      <a href="" class="pull-left">
        <img class="navbar-logo" src="../assets/logo/icon.svg" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	      
        <li><a href="../setup.html">Setup</a></li>

        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Lessons <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-unix/index.html">UNIX command line</a></li>
            
            
            <li><a href="../02-slurm/index.html">SLURM workload manager</a></li>
            
            
            <li><a href="../03-blast/index.html">BLAST search</a></li>
            
            
            <li><a href="../04-parse/index.html">Parsing BLAST output</a></li>
            
            
            <li><a href="../05-tree/index.html">Phylogenetic tree</a></li>
            
          </ul>
        </li>

        

      

      <li class="dropdown">
        <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Optional <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../10-more_unix/index.html">More UNIX commands</a></li>
          
            <li><a href="../20-text_process/index.html">FASTA manipulation with awk</a></li>
          
            <li><a href="../30-extra_tools/index.html">Another bioinformatic tool: PyANI</a></li>
          
            <li><a href="../40-awesome_bioinfo/index.html">Awesome bioinformatics</a></li>
          
            <li><a href="../50-multiple_gene_copies/index.html">Multiple gene copies</a></li>
          
            <li><a href="../60-bootstrapped_trees/index.html">ML trees with bootstraps</a></li>
          
        </ul>
      </li>

      

      <li class="dropdown">
      <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../about/">About</a></li>
          
            <li><a href="../acknowledgments/">Acknowledgements</a></li>
          
            <li><a href="../faq/index.html">FAQ</a></li>
          
            <li><a href="../intro_slides/index.html">Intro Slides</a></li>
          
        </ul>
      </li>

      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" title="Google search" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>




























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../04-parse/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Basic	bioinformatics and command-line	tools	for phytopathologists</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../"><span class="glyphicon glyphicon-menu-up" aria-hidden="true"></span><span class="sr-only">lesson home</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">Phylogenetic tree</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>












<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 30 min
      <br/>
      <strong>Exercises:</strong> 15 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can we run other software and create an analyses pipeline in the script?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Calling preinstalled software from the cluster computer</p>
</li>
	
	<li><p>Run a genomic analyses pipeline.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h2 id="sequence-alignment">Sequence alignment</h2>

<p><a href="https://en.wikipedia.org/wiki/Sequence_alignment">Sequence alignment</a> is the basic and the most important step in phylogenetic analysis. 
A wrong sequence alignment will lead to wrong result. 
Various alignment tools are used to identify similarity regions that indicate fuctional, 
structural, and/or evolutionary relationships sequences. 
Multiple sequence alignment tools such as ClustalOmega, Muscle, MAFFT are commonly used. 
They vary in terms of algorithms used; <a href="https://www.ebi.ac.uk/Tools/msa/clustalo/">ClustalOmega</a> uses HMM profile-profile techniques, 
and <a href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a> uses Fast Fourier Transforms and 
both are considered good for medium to large alignments. 
<a href="https://www.ebi.ac.uk/Tools/msa/muscle/">Muscle</a> is regarded good with proteins for medium alignments.</p>

<p>We will use MAFFT sequence alignment tool.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>module load mafft

<span class="gp">$ </span>mafft avrBs2_all_genomes.fas <span class="o">&gt;</span> avrBs2_all_genomes_aligned.fas
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>nthread = 0
generating a scoring matrix for nucleotide (dist=200) ... done
Gap Penalty = -1.53, +0.00, +0.00

Making a distance matrix ..
    1 / 20
done.
...
...
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>more avrBs2_all_genomes_aligned.fas
</code></pre></div></div>

<blockquote class="caution">
  <p>Make sure you are still in phylogeny folder. You can use <code class="language-plaintext highlighter-rouge">pwd</code> to check.</p>
</blockquote>

<blockquote class="tips">
  <p>Reminder: <code class="language-plaintext highlighter-rouge">more</code> displays small chunks of the file at a time. 
You can scroll down the file using <kbd>enter</kbd> key.
You can return to command line prompt by pressing <kbd>q</kbd> key.</p>
</blockquote>

<p>The two files avrBs2_all_genomes.fas and avrBs2_all_genomes_aligned.fas 
may not look very different at the moment considering the gene was present 
in all the sample genomes used here as seen from blast results. 
However, aligning the sequence prior to phylogenetic reconstruction is always valuable.
Now, we have an input file ready to use for a software that can run a 
<a href="https://www.ncbi.nlm.nih.gov/Class/NAWBIS/Modules/Phylogenetics/phylo15.html">maximum likelihood phylogentic analysis</a>. 
Read more about phylogenetic trees <a href="https://www.nature.com/scitable/topicpage/reading-a-phylogenetic-tree-the-meaning-of-41956/">here</a>.</p>

<h2 id="phylogenetic-tree">Phylogenetic tree</h2>

<p>There are several phylogenetic analyses software and tools that can be used. 
We will run a Maximum likelhood phylogenetic analysis. 
There are other models/methods such as Neighbor-Joining, Maximum parsimony, and Bayesian. 
However, our focus is on application of command line, 
we will only focus on running a specific analysis here.
Maximum Likelihood uses probabilistic values to model the evolution and 
the tree with the highest probability is shared.</p>

<p>We will use <a href="https://cme.h-its.org/exelixis/web/software/raxml/">RAxML - Randomized Axelerated Maximum Likelihood</a> tool to 
construct the tree that we will submit through a batch file.</p>

<p>The SLURM submission script is present in `/blue/general_workshop/share/scripts/slurm_tree.sh</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cp</span> ../../share/scripts/slurm_tree.sh ./

<span class="gp">$ </span><span class="nb">tail</span> <span class="nt">-n8</span> slurm_tree.sh
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Load RAxML
module load raxml

# Make tree with RAxML
raxmlHPC -d -p 12345 -m GTRGAMMAI -s avrBs2_all_genomes_aligned.fas -n avrBs2_tree

# Success message
echo 'ML tree output in RAxML_bestTree.avrBs2_tree'
</code></pre></div></div>

<p>The argument <code class="language-plaintext highlighter-rouge">-n</code> allows us to name the output suffix. In our case, the output will
be name <code class="language-plaintext highlighter-rouge">RAxML_bestTree.avrBs2_tree</code>.</p>

<blockquote class="tips">
  <h2 id="help-with-arguments">Help with arguments</h2>
  <p>To understand what other arguments mean, you can check the help file <code class="language-plaintext highlighter-rouge">ml raxml; raxmlHPC -h</code>.</p>
</blockquote>

<p>Edit the email address in the SLURM submission script and submit the job.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>nano slurm_tree.sh

<span class="gp">$ </span>sbatch slurm_tree.sh
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="running-bootstraps">Running bootstraps</h2>
  <p>To bootstrap the tree, add arguments <code class="language-plaintext highlighter-rouge">-b -#1000</code> 
where <code class="language-plaintext highlighter-rouge">1000</code> is the number of bootstraps.
If you run bootstraps, the output will be in a file named <code class="language-plaintext highlighter-rouge">RAxML_bootstrap.&lt;suffix&gt;</code>.
To understand about the arguments, you can run <code class="language-plaintext highlighter-rouge">ml raxml; raxmlHPC -h</code></p>
</blockquote>

<h2 id="tree-visualization">Tree visualization</h2>

<p>We can download this tree file in our own computer and 
visualize using ‘Figtree’.</p>

<p>Alternatively, we can use commandline-based utilities to generate image.
We’ll use <code class="language-plaintext highlighter-rouge">ggtree</code> package in <code class="language-plaintext highlighter-rouge">R</code> programming language.</p>

<p><code class="language-plaintext highlighter-rouge">R</code> and all necessary packages are already installed in Hipergator.
A helper R script is available from scripts directory.
Run the script as follows:</p>

<p><code class="language-plaintext highlighter-rouge">Rscript &lt;script&gt; &lt;input file&gt; &lt;output name&gt;</code></p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>ml R

<span class="gp">$ </span>Rscript ../../share/scripts/ggtree.R RAxML_bestTree.avrBs2_tree tree_avrbs2.pdf

<span class="gp">$ </span><span class="nb">ls </span>tree_avrbs2.pdf
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tree_avrbs2.pdf
</code></pre></div></div>

<p>Now we can navigate to OpenDemand dashboard to <strong>Files</strong> &gt; 
<strong>/blue/general_workshop</strong> &gt; <strong><code class="language-plaintext highlighter-rouge">username</code>/phylogeny/</strong>.
Select the file <strong>tree_avrbs2.pdf</strong> and click the <strong>Download</strong> button.</p>

<p><img src="/assets/img/tree.jpg" alt="Phylogenentic tree" /></p>

<blockquote class="challenge">
  <h2 id="optional-trees-in-ascii">(Optional) Trees in ASCII!</h2>
  <p>To quickly visualize the tree within commandline, try this:</p>
  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>ml newick_utils/1.6

<span class="gp">$ </span>nw_display RAxML_bestTree.avrBs2_tree
</code></pre></div>  </div>
</blockquote>

<h2 id="phylogenetic-tree-pipeline">Phylogenetic tree pipeline</h2>

<p>We have now successfully extracted a gene of interest from our genomes and 
generated a phylogenetic tree. 
However, we completed the entire process using several distinct steps where
we evaluated output of one step before starting to run the next step.
In real world scenario, most of the intermediate results are not of 
interest, so those disjoint steps are often chained
into one or few longer workflow, referred to as <strong>pipeline</strong>.</p>

<p>Question: Would it be possible to develop a script where we only need to 
provide external (generated outside the pipeline usch as gene and
genome sequences) inputs and get the final result (the phylogennetic tree) 
in return directly?</p>

<p>YES*</p>

<p>For this, we will merge all the steps together into one script.</p>

<p>Warning: *Chaining multiple steps is not always straightforward. 
We have to be vigilant for what might change when the steps are automated and 
intermediate results are not evaluated.
For example, we have to make sure that the genes of interest are actually in the genome. 
Otherwise, the BLAST may output match to different closely related gene, 
gene sequence alignment may add gaps and give us a wrong output and so on. 
Understanding what might go wrong comes from experience in bioinformatics.</p>

<p>Let’s put everything together into one script. The input files are
available in <code class="language-plaintext highlighter-rouge">/blue/general_workshop/share/all_in_one/</code> and the 
script is available as <code class="language-plaintext highlighter-rouge">/blue/general_workshop/share/scripts/slurm_pipeline.sh</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cd</span> /blue/general_workshop/&lt;username&gt;

<span class="gp">$ </span><span class="nb">mkdir </span>pipeline

<span class="gp">$ </span><span class="nb">cd </span>pipeline

<span class="gp">$ </span><span class="nb">cp</span> <span class="nt">-r</span> ../../share/phylogeny/<span class="k">*</span> ./

<span class="gp">$ </span><span class="nb">cp</span> ../../share/scripts/slurm_pipeline.sh ./

<span class="gp">$ </span><span class="nb">cp</span> ../../share/scripts/blast2fasta.sh ./

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>avrBs2.fas         GEV1054.fasta     GEV909.fasta     GEV968.fasta          Xeu.fasta
blast2fasta.sh     GEV1063.fasta     GEV915.fasta     GEV993.fasta          Xg.fasta
GEV1001.fasta      GEV839.fasta      GEV917.fasta     slurm_pipeline.sh     Xp.fasta
GEV1026.fasta      GEV893.fasta      GEV936.fasta     TB15.fasta
GEV1044.fasta      GEV904.fasta      GEV940.fasta     Xc.fasta
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">tail</span> <span class="nt">-n23</span> slurm_pipeline.sh
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Load programs
module load ncbi_blast
module load mafft
module load raxml

# Loop over each genome
for genome in `ls *.fasta | sed 's/.fasta//g'`
do
  # Make database
   makeblastdb -in "$genome.fasta" -dbtype nucl -out "$genome"

  # Run blastn on that database
   blastn -query avrBs2.fas -db "$genome" -out $genome"_avrBs2.out" -outfmt 5 -evalue 0.001
done

# Combine all outputs and parse into multiFASTA file
cat *_avrBs2.out | sh blast2fasta.sh &gt; avrBs2_all_genomes.fas

# Align sequences with MAAFT
mafft avrBs2_all_genomes.fas &gt; avrBs2_all_genomes_aligned.fas

# Make tree with RAxML
raxmlHPC -d -p 12345 -m GTRGAMMAI -s avrBs2_all_genomes_aligned.fas -n avrBs2_tree
</code></pre></div></div>

<p>After editing the email address in <code class="language-plaintext highlighter-rouge">nano</code>, you are ready to submit the submission script.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>sbatch slurm_pipeline.sh
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="visualize-tree">Visualize tree</h2>
  <p>Once the job finishes, you can transfer all the outputs to your personal computer. 
Then, you can visualize the ‘RAxML_bestTree.avrBs2_tree’ using Figtree software. 
<a href="/setup.html">Setup page</a> has instructions on how to install Figtree.</p>
</blockquote>





<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Don’t forget to check parameters of the software. Can be viewed using ‘-h’ followed by the software name.</p>
</li>
    
  </ul>
</blockquote>

</article>



























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../04-parse/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../"><span class="glyphicon glyphicon-menu-up" aria-hidden="true"></span><span class="sr-only">lesson home</span></a>
      
    </h3>
  </div>
</div>


      
      
<footer>
  <hr/>
  <div class="row">
    <div class="col-md-6" align="left">
      <h4>
	      Copyright &copy; 2020–2022 <a href="https://github.com/BioInfoAPS">APS Basic Bioinformatics Workshop Team</a>
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>

<script>
/* Based on https://stackoverflow.com/a/30894829/1802726 */
var show_instructions = function() {
     //set all nav tabs to inactive
    $('.nav-tabs li').removeClass('active');

    //get all nav tabs matching the os and set to active
    $('.nav-tabs li a[href*="'+localStorage.os+'"]').closest('li').addClass('active');

    //set all tabs to inactive
    $('.tab-pane').removeClass('active');

    //get all tabs matching the os and set to active
    $('.tab-pane[id*="'+localStorage.os+'"]').addClass('active');
}

if (typeof(Storage) !== "undefined") {
    if (localStorage.os == "undefined") {
        localStorage.os = "windows";
    }
    show_instructions();
} else {
    console.log("No localStorage");
}

/* Based on https://stackoverflow.com/a/30894829/1802726 */
$('.nav-tabs li a').click(function(){
    //get selected os
    localStorage.os = $(this)[0].dataset.os;
    show_instructions();
});

  // detect misssing iframe for Eventbrite
  $(document).ready(function(){
      if ($('iframe').height() == 0) {
          $('iframe').before('<p>Looks like your adblocker blocked the registration window. Please navigate to <a href="https://www.eventbrite.com/tickets-external?eid=&ref=etckt">https://www.eventbrite.com/tickets-external?eid=&ref=etckt</a> to register.</p>');
      }
  });

  // hide alerts when site is rendered on carpentries.github.io
  $(document).ready(function() {
      if (location.href.startsWith("https://carpentries.github.io/workshop-template/")) {
          $("div.alert").hide();
      }
  });

</script>

  </body>
</html>
