















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-10-17 14:45:52 -0400">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="https://bioinfoaps.github.io/">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    <link rel="license" href="#license-info" />

    

    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Basic	bioinformatics and command-line	tools	for phytopathologists"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  Parsing BLAST output &ndash; Basic	bioinformatics and command-line	tools	for phytopathologists
  </title>  

  </head>
  <body>

    <div class="container">
      
















  
  















<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      <a href="" class="pull-left">
        <img class="navbar-logo" src="../assets/logo/icon.svg" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	      
        <li><a href="../setup.html">Setup</a></li>

        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Lessons <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-unix/index.html">UNIX command line</a></li>
            
            
            <li><a href="../02-slurm/index.html">SLURM workload manager</a></li>
            
            
            <li><a href="../03-blast/index.html">BLAST search</a></li>
            
            
            <li><a href="../04-parse/index.html">Parsing BLAST output</a></li>
            
            
            <li><a href="../05-tree/index.html">Phylogenetic tree</a></li>
            
          </ul>
        </li>

        

      

      <li class="dropdown">
        <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Optional <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../10-more_unix/index.html">More UNIX commands</a></li>
          
            <li><a href="../20-text_process/index.html">FASTA manipulation with awk</a></li>
          
            <li><a href="../30-extra_tools/index.html">Another bioinformatic tool: PyANI</a></li>
          
            <li><a href="../40-awesome_bioinfo/index.html">Awesome bioinformatics</a></li>
          
            <li><a href="../50-multiple_gene_copies/index.html">Multiple gene copies</a></li>
          
        </ul>
      </li>

      

      <li class="dropdown">
      <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../about/">About</a></li>
          
            <li><a href="../acknowledgments/">Acknowledgements</a></li>
          
            <li><a href="../faq/index.html">FAQ</a></li>
          
            <li><a href="../intro_slides/index.html">Intro Slides</a></li>
          
        </ul>
      </li>

      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" title="Google search" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>




























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../03-blast/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Basic	bioinformatics and command-line	tools	for phytopathologists</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../05-tree/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">Parsing BLAST output</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>












<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 35 min
      <br/>
      <strong>Exercises:</strong> 15 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can we modify files in UNIX?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Modify file formats</p>
</li>
	
	<li><p>Running bash script</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h2 id="file-parsing">File parsing</h2>

<p>In a bioinformatics project, it is common to use several programs in series
i.e. output of a program is used as input to another program. 
There are many standard data formats such as 
<a href="https://en.wikipedia.org/wiki/FASTA_format" target="_blank">FASTA</a>, 
<a href="https://samtools.github.io/hts-specs/SAMv1.pdf" target="_blank">SAM and BAM</a> which are 
commonly accepted by most programs. 
This makes transition from one program to another program easier.
Unfortunately, sometimes programs output non-standard or 
proprietary formats, causing disconnect in using downstream programs. 
This is where file parsing is important. 
File parsing simply means reading the file and facilitating the data 
transformation for subsequent analysis.</p>

<p>Note: While file parsing does not sound important, a bioinformatician has 
to spend significant time correcting file formats.</p>

<h2 id="running-blastn-for-phylogenetic-tree">Running BLASTn for phylogenetic tree</h2>

<p>In next lesson, we will phylogenetically compare ‘avrBs2’ gene 
from various strains of <em>Xanthomonas</em> species.
The ouline of activity will be as follows:</p>

<p>BLAST → parse to fasta format → align sequence → phylogenetic analysis</p>

<p>To get started, we will run BLASTn in SLURM just like last section, 
however, we will work with a larger set (20) of <em>Xanthomonas</em> genomes.
You can find the genomes in the directory 
<code class="language-plaintext highlighter-rouge">/blue/general_workshop/share/phylogeny</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cd</span> /blue/general_workshop/&lt;username&gt;

<span class="gp">$ </span><span class="nb">cp</span> <span class="nt">-r</span> ../share/phylogeny ./

<span class="gp">$ </span><span class="nb">cd </span>phylogeny

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>avrBs2.fas     GEV1063.fasta     GEV915.fasta     GEV993.fasta     Xp.fasta
...
...
</code></pre></div></div>

<p>The SLURM submission script is located in <code class="language-plaintext highlighter-rouge">/blue/general_workshop/share/scripts/slurm_blast_xml.sh</code></p>

<blockquote class="caution">
  <p>This is not the same script as last section. Use this script, not the last one.</p>
</blockquote>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">tail</span> <span class="nt">-n9</span> ../../share/scripts/slurm_blast_xml.sh
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Loop over each genome
for genome in `ls *.fasta | sed 's/.fasta//g'`
do
  # Make database
  makeblastdb -in "$genome.fasta" -dbtype nucl -out "$genome"

  # Run blastn on that database
  blastn -query avrBs2.fas -db "$genome" -out $genome"_avrBs2.out" -𝗼𝘂𝘁𝗳𝗺𝘁 𝟱 -evalue 0.001
done
</code></pre></div></div>

<blockquote class="notes">
  <p>This script has a new argument <code class="language-plaintext highlighter-rouge">-outfmt 5</code> for <code class="language-plaintext highlighter-rouge">blastn</code> command. This will output the result in XML format.
The <code class="language-plaintext highlighter-rouge">.out</code> files you generated in previous section is not in XML format.</p>
</blockquote>

<blockquote class="challenge">
  <h2 id="exercise-submit-a-job-with-blast-script">Exercise: Submit a job with BLAST script</h2>

  <p>The obejctive of this exercise is to be able to run SLURM script<br />
You can use checklist to track progress.</p>
  <ol>
    <li>Copy script <code class="language-plaintext highlighter-rouge">slurm_blast_xml.sh</code> to your working directory (<code class="language-plaintext highlighter-rouge">phylogeny</code>). <input type="checkbox" /></li>
    <li>Edit email address in SLURM submission script. <input type="checkbox" /></li>
    <li>Submit the SLURM script. <input type="checkbox" /></li>
    <li>Make sure the all output files are present at the end of the job. <input type="checkbox" /></li>
  </ol>

  <details>
    <summary></summary>
    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#1</span>
<span class="gp">$ </span><span class="nb">cp</span> ../../share/scripts/slurm_blast_xml.sh ./

<span class="c">#2</span>
<span class="gp">$ </span>nano slurm_blast_xml.sh
→ edit email address → ctrl+x → y → press enter

<span class="c">#3</span>
<span class="gp">$ </span>sbatch slurm_blast_xml.sh

<span class="c">#4</span>
<span class="gp">$ </span><span class="nb">ls</span> <span class="k">*</span>_avrBs2.out | <span class="nb">wc</span> <span class="nt">-l</span>
</code></pre></div>    </div>

    <p>The output of answer #4 should be 20 if everything went fine.</p>

  </details>
</blockquote>

<h2 id="blast-xml-output-vs-fasta-format">BLAST XML output vs FASTA format</h2>

<p>Now, we can convert the BLAST <code class="language-plaintext highlighter-rouge">XML</code> output in a more common format
that can be read by multiple software to run analyses. 
Specifically, we will convert it to <code class="language-plaintext highlighter-rouge">FASTA</code> format, which we
will subsequently use to generate a phylogenetic tree.</p>

<p>BLAST XML format includes a lot of information on query and database name, 
sequence identity, statistics, alignments and so on.
Below, you can see what BLAST XML output looks like:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cat </span>Xeu_avrBs2.out
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE BlastOutput PUBLIC "-//NCBI//NCBI BlastOutput/EN" "http://www.ncbi.nlm.nih.gov/dtd/NCBI_BlastOutput.dtd"&gt;
&lt;BlastOutput&gt;
  &lt;BlastOutput_program&gt;blastn&lt;/BlastOutput_program&gt;
  &lt;BlastOutput_version&gt;BLASTN 2.10.1+&lt;/BlastOutput_version&gt;
  ...
  ...
  &lt;𝗕𝗹𝗮𝘀𝘁𝗢𝘂𝘁𝗽𝘂𝘁_𝗱𝗯&gt;𝗫𝗲𝘂&lt;/𝗕𝗹𝗮𝘀𝘁𝗢𝘂𝘁𝗽𝘂𝘁_𝗱𝗯&gt;
  ...
  ...
&lt;/BlastOutput&gt;
...
...
&lt;Hit&gt;
  &lt;Hit_num&gt;1&lt;/Hit_num&gt;
  &lt;Hit_hsps&gt;
    &lt;Hsp&gt;
      &lt;Hsp_bit-score&gt;4170.86&lt;/Hsp_bit-score&gt;
      &lt;Hsp_score&gt;2258&lt;/Hsp_score&gt;
      &lt;Hsp_evalue&gt;0&lt;/Hsp_evalue&gt;
      ...
      ...
      &lt;Hsp_qseq&gt;AAGTGCTGGCAACGCGTCCAAACACGAGCAGGCCAGGCAGACCGAGACGGATTGA&lt;/Hsp_qseq&gt;
      &lt;𝗛𝘀𝗽_𝗵𝘀𝗲𝗾&gt;𝗔𝗔𝗚𝗧𝗚𝗖𝗧𝗚𝗚𝗖𝗔𝗔𝗖𝗚𝗖𝗚𝗧𝗖𝗖𝗔𝗔𝗔𝗖𝗔𝗖𝗖𝗔𝗚𝗖𝗔𝗚𝗚𝗖𝗖𝗔𝗚𝗚𝗖𝗔𝗚𝗔𝗖𝗖𝗚𝗔𝗚𝗔𝗖𝗚𝗚𝗔𝗧𝗧𝗚𝗔&lt;/𝗛𝘀𝗽_𝗵𝘀𝗲𝗾&gt;
      &lt;Hsp_midline&gt;||||||||||||||||||||||||| |||||||||||||||||||||||||||||&lt;/Hsp_midline&gt;
    &lt;/Hsp&gt;
  &lt;/Hit_hsps&gt;
&lt;/Hit&gt;
</code></pre></div></div>

<blockquote class="notes">
  <p>The output above is for demonstration purpose and the real output may look different.</p>
</blockquote>

<p>We will have to convert it into FASTA format which only needs two line from the XML:
<code class="language-plaintext highlighter-rouge">BlastOutput_db</code> and <code class="language-plaintext highlighter-rouge">Hsp_hseq</code>. The FASTA format looks like:</p>

<div class="language-plaintext terminal highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt; Xeu
AAGTGCTGGCAACGCGTCCAAACACCAGCAGGCCAGGCAGACCGAGACGGATTGA
</code></pre></div></div>

<h3 id="bash-script-to-parse-xml-to-fasta">Bash script to parse XML to FASTA</h3>

<p>We will now use a bash script to parse the BLAST output to 
extract the file as a multiFASTA file.
The script is located in path <code class="language-plaintext highlighter-rouge">/blue/general_workshop/share/scripts/blast2fasta.sh</code></p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cp</span> ../../share/scripts/blast2fasta.sh ./   

<span class="gp">$ </span><span class="nb">cat </span>blast2fasta.sh
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>#!/bin/bash
while read line
do
        [[ `grep "BlastOutput_db" &lt;&lt;&lt; $line` ]] &amp;&amp; echo -n "&gt;" &lt;&lt;&lt; $line
        [[ `grep -E "BlastOutput_db|Hsp_hseq" &lt;&lt;&lt; $line` ]] &amp;&amp; sed -n "s:.*&gt;\(.*\)&lt;/.*:\1:p" &lt;&lt;&lt; $line
done
</code></pre></div></div>

<p>What this script does is loop line by line (<code class="language-plaintext highlighter-rouge">while</code>) over the file, 
and look for (<code class="language-plaintext highlighter-rouge">grep</code>) keywords <code class="language-plaintext highlighter-rouge">BlastOutput_db</code> and <code class="language-plaintext highlighter-rouge">Hsp_hseq</code>.
If it find the keywords, then it will extract (<code class="language-plaintext highlighter-rouge">sed</code>) the contents
between those tags and output in appropriate format.</p>

<blockquote class="tips">
  <p>Considering the course objective, we won’t cover the exact syntax of the script at this time.</p>
</blockquote>

<h3 id="running-a-bash-script">Running a bash script</h3>

<p>We can call the bash script in two ways</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">./&lt;script_name&gt;</code></li>
  <li><code class="language-plaintext highlighter-rouge">sh &lt;scriptname&gt;</code></li>
</ul>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cat</span> <span class="k">*</span>_avrBs2.out | ./blast2fasta.sh <span class="o">&gt;</span> avrBs2_all_genomes.fas
</code></pre></div></div>

<p>The FASTA output is stored in file <code class="language-plaintext highlighter-rouge">avrBs2_all_genomes.fas</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cat </span>avrBs2_all_genomes.fas
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;GEV1026
AAGTGCTGGCAACGCGTCCAAACAAGGCCTGCGCCGCACGCCTGCCAGCGCGCGCAACGCAGGCATCGTTTCGCATCCGGG
CGGTACTTTTCGCCTAATTTGCCAATTGTCATATGCCACGCGCTTTACTGGCCGCCCGCCGCGTTTTCGAGGTCATCATGC
GCATCGGTCCTCTGCAACCTTCTATCGCGCACACTGC
...
...
&gt;GEV1001
GTGCTGGCAACGCGTCCAAACAAGGCCTGCGCCGCACGCCTGCCAGCGCGCGTAACGCAGGCATCGTTTCGCATCGGCGGG
CGGTACTTTTCGCCTAATTTGCCAATTGTCAGATGCCACGCGCTTTACTGGCCGCCCGACGCGTTTTCGAGGTCATCATGC
GCATCGGTCCTCTGCAACCTTCTATCGCGCA
...
...
</code></pre></div></div>

<blockquote class="notes">
  <p>The output above is for demonstration purpose and the real output may look different.</p>
</blockquote>

<blockquote class="tips">
  <h2 id="when-to-use-slurm">When to use SLURM?</h2>
  <p>We are not using SLURM for this job because it is fast and not resource intensive.
Use SLURM for resource intensive or time-consuming tasks.</p>
</blockquote>

<p>Now we have a multiFASTA file with sequence id for each genome 
we used along with their respective sequence. 
We can now use this FASTA formatted file to run sequence alignment and 
subsequently construct a tree.</p>





<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>You can run bash command as a script in .sh file format.</p>
</li>
    
  </ul>
</blockquote>

</article>



























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../03-blast/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../05-tree/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>


      
      
<footer>
  <hr/>
  <div class="row">
    <div class="col-md-6" align="left">
      <h4>
	      Copyright &copy; 2020–2022 <a href="https://github.com/BioInfoAPS">APS Basic Bioinformatics Workshop Team</a>
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>

<script>
/* Based on https://stackoverflow.com/a/30894829/1802726 */
var show_instructions = function() {
     //set all nav tabs to inactive
    $('.nav-tabs li').removeClass('active');

    //get all nav tabs matching the os and set to active
    $('.nav-tabs li a[href*="'+localStorage.os+'"]').closest('li').addClass('active');

    //set all tabs to inactive
    $('.tab-pane').removeClass('active');

    //get all tabs matching the os and set to active
    $('.tab-pane[id*="'+localStorage.os+'"]').addClass('active');
}

if (typeof(Storage) !== "undefined") {
    if (localStorage.os == "undefined") {
        localStorage.os = "windows";
    }
    show_instructions();
} else {
    console.log("No localStorage");
}

/* Based on https://stackoverflow.com/a/30894829/1802726 */
$('.nav-tabs li a').click(function(){
    //get selected os
    localStorage.os = $(this)[0].dataset.os;
    show_instructions();
});

  // detect misssing iframe for Eventbrite
  $(document).ready(function(){
      if ($('iframe').height() == 0) {
          $('iframe').before('<p>Looks like your adblocker blocked the registration window. Please navigate to <a href="https://www.eventbrite.com/tickets-external?eid=&ref=etckt">https://www.eventbrite.com/tickets-external?eid=&ref=etckt</a> to register.</p>');
      }
  });

  // hide alerts when site is rendered on carpentries.github.io
  $(document).ready(function() {
      if (location.href.startsWith("https://carpentries.github.io/workshop-template/")) {
          $("div.alert").hide();
      }
  });

</script>

  </body>
</html>
