















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-09-21 16:05:38 -0400">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="https://bioinfoaps.github.io/">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    <link rel="license" href="#license-info" />

    

    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Basic	bioinformatics and command-line	tools	for phytopathologists"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  Another bioinformatic tool: PyANI &ndash; Basic	bioinformatics and command-line	tools	for phytopathologists
  </title>  

  </head>
  <body>

    <div class="container">
      
















  
  















<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      <a href="" class="pull-left">
        <img class="navbar-logo" src="../assets/logo/icon.svg" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	      
        <li><a href="../setup.html">Setup</a></li>

        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Lessons <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-unix/index.html">UNIX command line</a></li>
            
            
            <li><a href="../02-slurm/index.html">SLURM workload manager</a></li>
            
            
            <li><a href="../03-blast/index.html">BLAST search</a></li>
            
            
            <li><a href="../04-parse/index.html">Parsing BLAST output</a></li>
            
            
            <li><a href="../05-tree/index.html">Phylogenetic tree</a></li>
            
          </ul>
        </li>

        

      

      <li class="dropdown">
        <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Optional <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../10-more_unix/index.html">More UNIX commands</a></li>
          
            <li><a href="../20-text_process/index.html">FASTA manipulation with awk</a></li>
          
            <li><a href="../30-extra_tools/index.html">Another bioinformatic tool: PyANI</a></li>
          
            <li><a href="../40-awesome_bioinfo/index.html">Awesome bioinformatics</a></li>
          
            <li><a href="../50-multiple_gene_copies/index.html">Multiple gene copies</a></li>
          
        </ul>
      </li>

      

      <li class="dropdown">
      <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../about/">Authors</a></li>
          
            <li><a href="../acknowledgments/">Acknowledgements</a></li>
          
            <li><a href="../faq/index.html">FAQ</a></li>
          
            <li><a href="../intro_slides/index.html">Intro Slides</a></li>
          
        </ul>
      </li>

      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" title="Google search" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>




























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../20-text_process/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Basic	bioinformatics and command-line	tools	for phytopathologists</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../40-awesome_bioinfo/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">Another bioinformatic tool: PyANI</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>






<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>To calculate average nucleotide identity of a genome with related genomes.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>In the regular lessons, we implemented three bioinformatic tools: 
<code class="language-plaintext highlighter-rouge">blast</code> for homology search,
<code class="language-plaintext highlighter-rouge">maaft</code> for sequence alignment, and
<code class="language-plaintext highlighter-rouge">raxml</code> for phylogenetic analysis.</p>

<p>In this section, we will discuss two more tools.</p>

<h2 id="pyani">PyANI</h2>
<p>PyANI is an open-source python-based tool for calculating 
Average Nucleotide Identity (ANI) between two or more sequences.
When comparing two genomes, first syntenic regions are identified
using tools such as <code class="language-plaintext highlighter-rouge">mummer</code> or <code class="language-plaintext highlighter-rouge">blast</code>.
Then the nucleotide identity is calculated in the syntenic regions.</p>

<p>The source code for <strong>PyANI</strong> is available at 
<a href="https://github.com/widdowquinn/pyani" target="_blank">widdowquinn/pyani</a>.
The documentation for basic usage is available 
<a href="https://github.com/widdowquinn/pyani/blob/master/README_v_0_2_x.md" target="_blank">here</a>.</p>

<p>The first stage of PyANI is</p>

<p><strong>PyANI</strong> v2 is available in Hipergator, but has to be loaded.
The dependencies, <code class="language-plaintext highlighter-rouge">mummer</code> and <code class="language-plaintext highlighter-rouge">blast+</code> will be loaded together with <code class="language-plaintext highlighter-rouge">pyani</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>ml pyani
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Lmod is automatically replacing "python/3.8" with "pyani/0.2.10".
</code></pre></div></div>

<p>We will be using the genomes present in <code class="language-plaintext highlighter-rouge">files/ani</code> for computing ANI.
The file <code class="language-plaintext highlighter-rouge">UXhortspp.fasta</code> contains genome of a unknown <em>X. hortorum</em> species.
The other sequences are genome of some <em>X. hortorum</em> pathovars 
downloaded from NCBI.</p>

<blockquote class="tips">
  <h2 id="getting-genome-sequences-from-ncbi">Getting genome sequences from NCBI</h2>
  <p>PyANI has a script called <code class="language-plaintext highlighter-rouge">genbank_get_genomes_by_taxon.py</code> to download 
all genomes for a taxon from NCBI.
For usage, check the documentation linked above.</p>
</blockquote>

<p>The objective now is to perform pairwise comparisons of all reference genomes
and calculate ANI. This can be performed with following command.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>average_nucleotide_identity.py <span class="nt">-i</span> files/ani <span class="nt">-o</span> ani <span class="nt">-m</span> ANIm <span class="nt">-g</span> <span class="nt">--gformat</span> png,pdf
</code></pre></div></div>

<blockquote class="notes">
  <ul>
    <li><code class="language-plaintext highlighter-rouge">average_nucleotide_identity.py</code> is the name of the script</li>
    <li><code class="language-plaintext highlighter-rouge">-i</code> is used to specify directory containing input genomes/sequences.</li>
    <li><code class="language-plaintext highlighter-rouge">-o</code> is used to specify output directory.
Note that the program will exit if this directory preexists.</li>
    <li><code class="language-plaintext highlighter-rouge">-m</code> is used to specify mode for alignment of syntenic region. 
<code class="language-plaintext highlighter-rouge">ANIm</code> specifies <code class="language-plaintext highlighter-rouge">mummer</code> and <code class="language-plaintext highlighter-rouge">ANIb</code> specifies <code class="language-plaintext highlighter-rouge">blast+</code>.</li>
    <li><code class="language-plaintext highlighter-rouge">-g</code> is used to generate graphic output, i.e., heatmap.
<code class="language-plaintext highlighter-rouge">--gformat</code> specifies the graphic output formats.</li>
  </ul>
</blockquote>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">ls </span>ani
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ANIm_alignment_coverage.pdf  ANIm_hadamard.pdf             ANIm_similarity_errors.pdf
ANIm_alignment_coverage.png  ANIm_hadamard.png             ANIm_similarity_errors.png
ANIm_alignment_coverage.tab  ANIm_hadamard.tab             ANIm_similarity_errors.tab
ANIm_alignment_lengths.pdf   ANIm_percentage_identity.pdf  nucmer_output.tar.gz
ANIm_alignment_lengths.png   ANIm_percentage_identity.png
ANIm_alignment_lengths.tab   ANIm_percentage_identity.tab
</code></pre></div></div>

<p>You can now transfer <code class="language-plaintext highlighter-rouge">ANIm_percentage_identity.png</code> 
to your computer to view the heatmap.</p>

<p><img src="/assets/img/ANIm_percentage_identity.png" height="500px" /></p>

<p>You can also get numeric ANI values from the <code class="language-plaintext highlighter-rouge">ANIm_percentage_identity.tab</code> file.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">awk</span> <span class="s1">'NR==1{print "-"$0; next}{for (i=2; i&lt;=NF; i++) {$i=substr($i,1,5)}; print $0}'</span> ani8/ANIm_percentage_identity.tab | column <span class="nt">-t</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>-             Xpopuli  Xhgardneri  Xhcynarae  Xhhederae  Xhvitians  Xhunknown  Xhcarotae  Xhtaraxaci  Xhpelargonii
Xpopuli       1.0      0.913       0.913      0.912      0.912      0.913      0.913      0.912       0.913
Xhgardneri    0.913    1.0         0.993      0.960      0.982      0.999      0.961      0.974       0.958
Xhcynarae     0.913    0.993       1.0        0.960      0.984      0.993      0.961      0.975       0.958
Xhhederae     0.912    0.960       0.960      1.0        0.960      0.960      0.965      0.955       0.961
Xhvitians     0.912    0.982       0.984      0.960      1.0        0.982      0.961      0.976       0.958
Xhunknown     0.913    0.999       0.993      0.960      0.982      1.0        0.961      0.974       0.958
Xhcarotae     0.913    0.961       0.961      0.965      0.961      0.961      1.0        0.956       0.961
Xhtaraxaci    0.912    0.974       0.975      0.955      0.976      0.974      0.956      1.0         0.954
Xhpelargonii  0.913    0.958       0.958      0.961      0.958      0.958      0.961      0.954       1.0
</code></pre></div></div>

<p>Based on ANI, the unknown strain seems closest to <em>X. hortorum pv. gardneri</em>.</p>

</article>



























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../20-text_process/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../40-awesome_bioinfo/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>


      
      
<footer>
  <hr/>
  <div class="row">
    <div class="col-md-6" align="left">
      <h4>
	      Copyright &copy; 2020â€“2022 <a href="https://github.com/BioInfoAPS">APS Basic Bioinformatics Workshop Team</a>
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>

<script>
/* Based on https://stackoverflow.com/a/30894829/1802726 */
var show_instructions = function() {
     //set all nav tabs to inactive
    $('.nav-tabs li').removeClass('active');

    //get all nav tabs matching the os and set to active
    $('.nav-tabs li a[href*="'+localStorage.os+'"]').closest('li').addClass('active');

    //set all tabs to inactive
    $('.tab-pane').removeClass('active');

    //get all tabs matching the os and set to active
    $('.tab-pane[id*="'+localStorage.os+'"]').addClass('active');
}

if (typeof(Storage) !== "undefined") {
    if (localStorage.os == "undefined") {
        localStorage.os = "windows";
    }
    show_instructions();
} else {
    console.log("No localStorage");
}

/* Based on https://stackoverflow.com/a/30894829/1802726 */
$('.nav-tabs li a').click(function(){
    //get selected os
    localStorage.os = $(this)[0].dataset.os;
    show_instructions();
});

  // detect misssing iframe for Eventbrite
  $(document).ready(function(){
      if ($('iframe').height() == 0) {
          $('iframe').before('<p>Looks like your adblocker blocked the registration window. Please navigate to <a href="https://www.eventbrite.com/tickets-external?eid=&ref=etckt">https://www.eventbrite.com/tickets-external?eid=&ref=etckt</a> to register.</p>');
      }
  });

  // hide alerts when site is rendered on carpentries.github.io
  $(document).ready(function() {
      if (location.href.startsWith("https://carpentries.github.io/workshop-template/")) {
          $("div.alert").hide();
      }
  });

</script>

  </body>
</html>
