















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-10-17 14:45:52 -0400">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="https://bioinfoaps.github.io/">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    <link rel="license" href="#license-info" />

    

    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Basic	bioinformatics and command-line	tools	for phytopathologists"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  UNIX command line &ndash; Basic	bioinformatics and command-line	tools	for phytopathologists
  </title>  

  </head>
  <body>

    <div class="container">
      
















  
  















<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      <a href="" class="pull-left">
        <img class="navbar-logo" src="../assets/logo/icon.svg" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	      
        <li><a href="../setup.html">Setup</a></li>

        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Lessons <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-unix/index.html">UNIX command line</a></li>
            
            
            <li><a href="../02-slurm/index.html">SLURM workload manager</a></li>
            
            
            <li><a href="../03-blast/index.html">BLAST search</a></li>
            
            
            <li><a href="../04-parse/index.html">Parsing BLAST output</a></li>
            
            
            <li><a href="../05-tree/index.html">Phylogenetic tree</a></li>
            
          </ul>
        </li>

        

      

      <li class="dropdown">
        <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Optional <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../10-more_unix/index.html">More UNIX commands</a></li>
          
            <li><a href="../20-text_process/index.html">FASTA manipulation with awk</a></li>
          
            <li><a href="../30-extra_tools/index.html">Another bioinformatic tool: PyANI</a></li>
          
            <li><a href="../40-awesome_bioinfo/index.html">Awesome bioinformatics</a></li>
          
            <li><a href="../50-multiple_gene_copies/index.html">Multiple gene copies</a></li>
          
        </ul>
      </li>

      

      <li class="dropdown">
      <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../about/">About</a></li>
          
            <li><a href="../acknowledgments/">Acknowledgements</a></li>
          
            <li><a href="../faq/index.html">FAQ</a></li>
          
            <li><a href="../intro_slides/index.html">Intro Slides</a></li>
          
        </ul>
      </li>

      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" title="Google search" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>




























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../"><span class="glyphicon glyphicon-menu-up" aria-hidden="true"></span><span class="sr-only">lesson home</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Basic	bioinformatics and command-line	tools	for phytopathologists</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../02-slurm/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">UNIX command line</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>












<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 20 min
      <br/>
      <strong>Exercises:</strong> 10 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can we navigate through folder and files in the computer using a command-line interface?</p>
</li>
	
	<li><p>What are the UNIX-commands to handle and edit files?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Learn how to navigate, open, and handle files using the terminal.</p>
</li>
	
	<li><p>Understand the basic commands to parse text files containing biological data.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>Command line interface and graphical user interaface are different ways of communicating with computer’s operating system. The shell is a program that provides the command line interface and allows to control the computer using keyboard commands. For bioinformatics tools, limited software have graphical user interface and you will have to use shell. The shell is a powerful method of communicating with the computer that can help you to do your work more efficienty and understanding how to use shell will be transformative for you to apply in bioinformatics. It can be used to connect to remote and cloud computers.</p>

<h2 id="terminal-command-line">Terminal Command line</h2>

<p>Once you login to HiperGator through SSH, you will start using a bash shell.</p>

<div class="language-plaintext terminal highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[&lt;username&gt;@login1 ~]$
</code></pre></div></div>

<p>The <code class="language-plaintext highlighter-rouge">$</code> prompt shows that the bash shell is ready to accept bash commands.</p>

<p>Before learning some basic commands, there are a few recommendations regarding UNIX systems.</p>

<ul>
  <li>The terminal syntax hates spaces between names. 
For long or complex names, use connectors such “_” or ‘-’ instead of spaces.</li>
  <li>Uppercase is different from lowercase. 
‘R’ is not the same as ‘r’ in commands, paths and arguments.</li>
  <li>UNIX system uses <code class="language-plaintext highlighter-rouge">/</code> for path, unlike windows, which uses <code class="language-plaintext highlighter-rouge">\</code>.</li>
</ul>

<p><img src="/assets/img/filesystem.svg" alt="UNIX filesystem" /></p>

<p>We will be working on ‘blue’ storage in HiperGator 
for our workshop under a group name ‘general_workshop’. 
Each user has a directory in the ‘general_workshop’ folder. 
There is also a ‘share’ directory where all the datasets and 
information for this workshop are stored. 
Please remember to only copy requested files from 
shared folder to your user folder and run the analyses 
in the folder(directory) with your username only.</p>

<p>Your personal folder is named as your gatorlink username. 
Enter the following command to go to your work directory 
(we will talk about <code class="language-plaintext highlighter-rouge">cd</code> shortly). 
Do not forget to replace &lt;username&gt; with the username provided to you.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cd</span> /blue/general_workshop/
</code></pre></div></div>

<blockquote class="caution">
  <p>When copying code, do not copy the <code class="language-plaintext highlighter-rouge">$</code> or <code class="language-plaintext highlighter-rouge">&gt;</code> prompt signs.
Selection of prompt is disabled in <strong>this</strong> website.</p>
</blockquote>

<h2 id="basic-commands">Basic Commands</h2>

<h3 id="displaying-current-pathlocation">Displaying current path/location</h3>

<p><code class="language-plaintext highlighter-rouge">pwd</code> displays your “path” (where you are located in the cluster).</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">pwd</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/blue/general_workshop/
</code></pre></div></div>

<h3 id="displaying-files-and-folders-in-current-location">Displaying files and folders in current location</h3>

<p><code class="language-plaintext highlighter-rouge">ls</code> command dipslays the files and folders in the current location.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>anujsharma     guest.11240     guest.11248     Intro_slides.pptx     share
emgoss         guest.11241     guest.11249     jhuguet               emgoss
...
...
</code></pre></div></div>

<p>Adding argument/flag <code class="language-plaintext highlighter-rouge">-l</code> to <code class="language-plaintext highlighter-rouge">ls</code> displays additional details such as permissions, file owner, 
size, date modified etc.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">ls</span> <span class="nt">-l</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>drwxr-sr-x 3 anujsharma      general_workshop    4096 Sep 10 03:47 anujsharma
...
...
-rw-r----- 1 jhuguet general_workshop 1290989 Sep  8 15:11 Intro_slides.pptx
drwxr-sr-x 7 jhuguet general_workshop    4096 Sep  8 15:27 share
drwxr-sr-x 2 emogss  general_workshop    4096 Sep  8 11:01 emgoss
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="permissions-in-linux">Permissions in linux</h2>

  <p><img src="/assets/img/permissions.svg" alt="Permissions in UNIX" /></p>

  <p>Linux permissions look like this: <code class="language-plaintext highlighter-rouge">d  r w x  r w −  r − −</code></p>

  <ul>
    <li>1st character represents special flag: file <code class="language-plaintext highlighter-rouge">−</code>, directory <code class="language-plaintext highlighter-rouge">d</code> or link <code class="language-plaintext highlighter-rouge">l</code>.</li>
    <li>The rest of the characters show permissions in set of three: <code class="language-plaintext highlighter-rouge">r</code> for read, <code class="language-plaintext highlighter-rouge">w</code> for write and <code class="language-plaintext highlighter-rouge">x</code> for execute. <code class="language-plaintext highlighter-rouge">−</code> means permission denied.
      <ul>
        <li>2nd to 4th characters: permission for file owner</li>
        <li>5th to 7th characters: permission for the group</li>
        <li>8th to 10th characters: permission for others</li>
      </ul>
    </li>
  </ul>
</blockquote>

<h3 id="creating-directories">Creating directories</h3>

<p><code class="language-plaintext highlighter-rouge">mkdir</code> creates a new directory in the current path.
Lets create a new directory called <code class="language-plaintext highlighter-rouge">newdir</code> and 
then use <code class="language-plaintext highlighter-rouge">ls</code> command to check if the directory was succesfully created.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cd</span> &lt;username&gt;

<span class="gp">$ </span><span class="nb">mkdir </span>newdir

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<blockquote class="notes">
  <p><code class="language-plaintext highlighter-rouge">cd &lt;username&gt;</code> is for entering your working directory first. We will cover <code class="language-plaintext highlighter-rouge">cd</code> shortly.</p>
</blockquote>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>newdir
</code></pre></div></div>

<h3 id="changing-directories">Changing directories</h3>

<p><code class="language-plaintext highlighter-rouge">cd</code> changes your current path. 
Lets change current path to the directory we just created. 
Use <code class="language-plaintext highlighter-rouge">pwd</code> to check the current path.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cd </span>newdir

<span class="gp">$ </span><span class="nb">pwd</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/blue/general_workshop/&lt;username&gt;/newdir
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cd</span> ..

<span class="gp">$ </span><span class="nb">pwd</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/blue/general_workshop/&lt;username&gt;
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="common-path-symbols-in-linux">Common path symbols in linux</h2>
  <p>Linux uses some symbols to represent commonly used paths.</p>
  <ul>
    <li><code class="language-plaintext highlighter-rouge">..</code> stands for parent directory.</li>
    <li><code class="language-plaintext highlighter-rouge">.</code> stands for current directory.</li>
    <li><code class="language-plaintext highlighter-rouge">/</code> at the beginning stands for root directory.</li>
    <li><code class="language-plaintext highlighter-rouge">~</code> stands for home directory.</li>
  </ul>
</blockquote>

<h3 id="copying-files">Copying files</h3>

<p><code class="language-plaintext highlighter-rouge">cp</code> is used for copying files. Lets copy <code class="language-plaintext highlighter-rouge">file1.txt</code> from share folder to your working directory.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cp</span> /blue/general_workshop/share/file1.txt ./file1.txt

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>file1.txt
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cp </span>file1.txt file2.txt

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>newdir     file1.txt     file2.txt
</code></pre></div></div>

<p><code class="language-plaintext highlighter-rouge">cp -r</code> can be used for copying entire directories. <code class="language-plaintext highlighter-rouge">-r</code> stands for recursive.
Copy the <code class="language-plaintext highlighter-rouge">demo</code> folder from share folder to your working directory.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cp</span> <span class="nt">-r</span> /blue/general_workshop/share/demo ./

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>demo     newdir     file1.txt     file2.txt
</code></pre></div></div>

<h3 id="moving-files">Moving files</h3>

<p><code class="language-plaintext highlighter-rouge">mv</code> is used for moving files or directories. Unlike copying, moving deletes the original copy.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">mv </span>file2.txt newfile.txt

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>demo     newdir     file1.txt     newfile.txt
</code></pre></div></div>

<h3 id="deleting-files">Deleting files</h3>

<p><code class="language-plaintext highlighter-rouge">rm</code> can be used for deleting files (and directories too with <code class="language-plaintext highlighter-rouge">-r</code> recursive argument)</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">rm </span>newfile.txt

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>demo     newdir     file1.txt
</code></pre></div></div>

<h3 id="removing-directories">Removing directories</h3>

<p><code class="language-plaintext highlighter-rouge">rmdir</code> removes the specified directory <strong>if it is empty</strong>. 
Lets remove the <code class="language-plaintext highlighter-rouge">newdir</code> we created earlier and 
check if it is removed using <code class="language-plaintext highlighter-rouge">ls</code>.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">rmdir </span>newdir

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>demo     file1.txt
</code></pre></div></div>

<h3 id="file-content-handling">File content handling</h3>

<p>There are a set of commands to read the contents of a file.</p>

<p><code class="language-plaintext highlighter-rouge">cat</code> reads the entire content of a file and returns to command line prompt. Let’s see what <code class="language-plaintext highlighter-rouge">file1.txt</code> file contains.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cat </span>file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008465.1      4979077 A       C       intergenic_region
CM008458.1      97095206        A       G       intergenic_region
CM008459.1      72668492        G       A       intergenic_region
...
...
CM008463.1      16380489        T       A       intergenic_region
CM008463.1      123496326       T       A       intergenic_region
CM008457.1      21333612        G       T       intergenic_region
</code></pre></div></div>

<p><code class="language-plaintext highlighter-rouge">less</code> and <code class="language-plaintext highlighter-rouge">more</code> display small chunks of the file content at a time.
In <code class="language-plaintext highlighter-rouge">less</code> output, you can scroll up and down the file content using 
<kbd>&uparrow;</kbd> and <kbd>&downarrow;</kbd> keys.
In <code class="language-plaintext highlighter-rouge">more</code> output, you can scroll down the file using <kbd>enter</kbd> key.
You can return to command line prompt by pressing <kbd>q</kbd> key.
Lets view <code class="language-plaintext highlighter-rouge">file1.txt</code> again but using <code class="language-plaintext highlighter-rouge">more</code> or <code class="language-plaintext highlighter-rouge">less</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>less file1.txt

<span class="gp">$ </span>more file1.txt
</code></pre></div></div>

<p><code class="language-plaintext highlighter-rouge">head</code> and <code class="language-plaintext highlighter-rouge">tail</code> can be used to read the start and end of the file respectively.
<code class="language-plaintext highlighter-rouge">-n</code> argument can be used to specify the the number of lines to read (default is 10 lines).
We can now only read the start or the end of <code class="language-plaintext highlighter-rouge">file1.txt</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">head </span>file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008465.1      4979077 A       C       intergenic_region
CM008458.1      97095206        A       G       intergenic_region
CM008459.1      72668492        G       A       intergenic_region
CM008465.1      57718962        G       A       downstream_gene_variant
CM008465.1      225524501       C       T       intergenic_region
CM008464.1      76483552        T       A       intergenic_region
CM008464.1      22690967        C       T       intergenic_region
CM008463.1      130960788       A       G       intergenic_region
CM008457.1      19966217        A       C       intergenic_region
CM008466.1      61441376        G       T       intergenic_region
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">tail</span> <span class="nt">-n</span> 3 file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008463.1      16380489        T       A       intergenic_region
CM008463.1      123496326       T       A       intergenic_region
CM008457.1      21333612        G       T       intergenic_region
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="extracting-lines-from-the-middle">Extracting lines from the middle</h2>
  <p>Lines from middle of a file can be extracted usng <code class="language-plaintext highlighter-rouge">sed -n</code> as follows. To extract 5th to 7th lines:</p>
  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">sed</span> <span class="nt">-n</span> 5,7p file1.txt
</code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008465.1      225524501       C       T       intergenic_region
CM008464.1      76483552        T       A       intergenic_region
CM008464.1      22690967        C       T       intergenic_region
</code></pre></div>  </div>

  <p><code class="language-plaintext highlighter-rouge">sed</code> is a very powerful tool in in bash and 
can be used to do wide range of text editing tasks.
Here, <code class="language-plaintext highlighter-rouge">-n</code> argument directs <code class="language-plaintext highlighter-rouge">sed</code> to pick only the lines that 
match the parameter <code class="language-plaintext highlighter-rouge">5,7p</code>.
You will see more uses of <code class="language-plaintext highlighter-rouge">sed</code> later.</p>
</blockquote>

<h3 id="file-length">File length</h3>

<p><code class="language-plaintext highlighter-rouge">wc</code> can be used for reading length of the file. 
<code class="language-plaintext highlighter-rouge">wc</code> accepts argument 
<code class="language-plaintext highlighter-rouge">-l</code> for number of lines, 
<code class="language-plaintext highlighter-rouge">-c</code> for number of characters and 
<code class="language-plaintext highlighter-rouge">-w</code> for number of words.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">wc</span> <span class="nt">-l</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>100 file1.txt
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">wc</span> <span class="nt">-c</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>4302 file1.txt
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">wc</span> <span class="nt">-w</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>500 file1.txt
</code></pre></div></div>

<h3 id="writing-a-file">Writing a file</h3>

<p><code class="language-plaintext highlighter-rouge">&gt;</code> operator can be used to write the output into a file.
Lets write some files.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">head</span> <span class="nt">-n</span> 5 file1.txt <span class="o">&gt;</span> head.txt

<span class="gp">$ </span><span class="nb">tail</span> <span class="nt">-n</span> 5 file1.txt <span class="o">&gt;</span> tail.txt

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>demo     file1.txt     head.txt     tail.txt
</code></pre></div></div>

<p>You can even save the lists of files in current directory into a file.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">ls</span> <span class="o">&gt;</span> files.txt

<span class="gp">$ </span><span class="nb">cat </span>files.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>demo
file1.txt
files.txt
head.txt
tail.txt
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="-vs-"><code class="language-plaintext highlighter-rouge">&gt;</code> vs <code class="language-plaintext highlighter-rouge">&gt;&gt;</code></h2>
  <p>Writing to an existing file with <code class="language-plaintext highlighter-rouge">&gt;</code> removes its existing contents. Use <code class="language-plaintext highlighter-rouge">&gt;&gt;</code> operator to append new contents to the end of exsiting file.</p>
</blockquote>

<h3 id="concatenate-files">Concatenate files</h3>

<p><code class="language-plaintext highlighter-rouge">cat</code> command is used for concatenation of multiple files. 
Lets concatenate the new files created in previous step. 
We can verify concatenation by checking the number of lines in concatenated file.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cat </span>head.txt tail.txt <span class="o">&gt;</span> concat.txt

<span class="gp">$ </span><span class="nb">wc</span> <span class="nt">-l</span> concat.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>10 concat.txt
</code></pre></div></div>

<blockquote class="notes">
  <p><code class="language-plaintext highlighter-rouge">head.txt</code>and <code class="language-plaintext highlighter-rouge">tail.txt</code> each had 5 lines each. To confirm, you can check like this: <code class="language-plaintext highlighter-rouge">wc -l head.txt</code>.</p>
</blockquote>

<h3 id="printoutput-to-screen">Print/output to screen</h3>

<p><code class="language-plaintext highlighter-rouge">echo</code> command can be used to display result to the screen.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">echo</span> <span class="s2">"hi"</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>hi
</code></pre></div></div>

<h3 id="variables">Variables</h3>

<p>Variables can be assigned values with <code class="language-plaintext highlighter-rouge">=</code> operator. Do not use space around <code class="language-plaintext highlighter-rouge">=</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nv">a</span><span class="o">=</span><span class="s2">"Hello"</span>

<span class="gp">$ </span><span class="nb">echo</span> <span class="nv">$a</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Hello
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nv">b</span><span class="o">=</span><span class="s2">"World"</span>

<span class="gp">$ </span><span class="nb">echo</span> <span class="s2">"</span><span class="nv">$a</span><span class="s2"> </span><span class="nv">$b</span><span class="s2">"</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Hello World
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="-vs--double-quotes-vs-single-quotes"><code class="language-plaintext highlighter-rouge">"</code> vs <code class="language-plaintext highlighter-rouge">'</code> (Double quotes vs single quotes)</h2>
  <p><code class="language-plaintext highlighter-rouge">"</code> and <code class="language-plaintext highlighter-rouge">'</code> mean different things in unix and should not be use interchangeably.
Check yourself how the output of <code class="language-plaintext highlighter-rouge">echo '$a $b'</code> differs from that of <code class="language-plaintext highlighter-rouge">echo "$a $b"</code>.</p>
</blockquote>

<h3 id="text-manipulation">Text manipulation</h3>

<p><code class="language-plaintext highlighter-rouge">cut</code> is used to extract fields of data from a string or a file.</p>

<p><code class="language-plaintext highlighter-rouge">head.txt</code> contains part of the output of a real sequence analysis. 
The first part (CM00084xx) contains the names of the chromosomes, 
which we need to extract.</p>

<p>Since chromosome name is first 8 characters, 
<code class="language-plaintext highlighter-rouge">cut -c</code> can be used to extract chromosome name. 
<code class="language-plaintext highlighter-rouge">-c</code> specifies vertical position of characters to extract</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cut</span> <span class="nt">-c</span> 1-8 head.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008465
CM008458
CM008459
CM008465
CM008465
</code></pre></div></div>

<p>The columns in <code class="language-plaintext highlighter-rouge">head.txt</code> are tab separated. You can verify with <code class="language-plaintext highlighter-rouge">cat head.txt</code>.
So, we can alternatively just extract the first <strong>column</strong> using <code class="language-plaintext highlighter-rouge">cut</code>.
<code class="language-plaintext highlighter-rouge">-f</code>: specify columns to extract.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cut</span> <span class="nt">-f</span> 1 head.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008465.1
CM008458.1
CM008459.1
CM008465.1
CM008465.1
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="cut-field-delimiter">Cut field delimiter</h2>

  <p>What if we want to define column by character other than <code class="language-plaintext highlighter-rouge">Tab</code>?
<code class="language-plaintext highlighter-rouge">-d</code> argument allows us to specify delimiter to break line into columns.
For example, suppose we don’t want the <code class="language-plaintext highlighter-rouge">.1</code> part in the chromosome name in the code above. 
In this case, we can get the chromosome name by separating columns 
by <code class="language-plaintext highlighter-rouge">.</code> instead of <code class="language-plaintext highlighter-rouge">Tab</code> and extracting the first column.</p>
  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cut</span> <span class="nt">-f</span> 1 <span class="nt">-d</span> <span class="s2">"."</span> head.txt
</code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008465
CM008458
CM008459
CM008465
CM008465
</code></pre></div>  </div>
</blockquote>

<h3 id="sorting">Sorting</h3>

<p><code class="language-plaintext highlighter-rouge">sort</code> is used to sort lines in a file. 
By default, <code class="language-plaintext highlighter-rouge">sort</code> sorts in alphanumeric order.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">sort </span>file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM0084󠁴̲5󠁴̲5.1      󠁴̲2󠁴̲7261226        C       A       intergenic_region
CM008󠁴4󠁴̲5󠁴̲5.1      󠁴̲2󠁴̲8󠁴̲9󠁴̲184259       AT      A       intron_variant
CM008󠁴4󠁴̲5󠁴̲5.1      󠁴̲2󠁴̲8󠁴̲9󠁴̲401193       G       A       intergenic_region
...
...
CM0084󠁴̲6󠁴̲6.1      󠁴̲61600177        C       T       intergenic_region
CM0084󠁴̲6󠁴̲6.1      󠁴̲71138968        G       A       intergenic_region
CM0084󠁴̲6󠁴̲6.1      󠁴̲95058324        T       C       intergenic_region
</code></pre></div></div>

<blockquote class="notes">
  <p>The top lines now begin with CM0084<strong>55</strong>.1 and the bottom ones with CM0084<strong>66</strong>.1.</p>
</blockquote>

<p><code class="language-plaintext highlighter-rouge">-r</code> argument reverses the order of sorting.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">sort</span> <span class="nt">-r</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM0084󠁴̲6󠁴̲6.1      󠁴̲95058324        T       C       intergenic_region
CM0084󠁴̲6󠁴̲6.1      󠁴̲71138968        G       A       intergenic_region
CM0084󠁴̲6󠁴̲6.1      󠁴̲61600177        C       T       intergenic_region
...
...
CM0084󠁴̲5󠁴̲5.1      󠁴̲2󠁴̲8󠁴̲9󠁴̲401193       G       A       intergenic_region
CM0084󠁴̲5󠁴̲5.1      󠁴̲2󠁴̲8󠁴̲9󠁴̲184259       AT      A       intron_variant
CM0084󠁴̲5󠁴̲5.1      󠁴̲2󠁴̲7󠁴̲261226        C       A       intergenic_region
</code></pre></div></div>

<p><code class="language-plaintext highlighter-rouge">-k</code> can be used to sort by a specific column.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">sort</span> <span class="nt">-k2</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008465.1      󠁴̲108474056       G       A       intergenic_region
CM008463.1      󠁴̲109077809       C       T       intergenic_region
...
...
CM008458.1      󠁴̲97095206        A       G       intergenic_region
CM008461.1      󠁴̲9871059 T       C       intergenic_region
</code></pre></div></div>

<p><code class="language-plaintext highlighter-rouge">-n</code> argument is used to sort in numerical ascending order.</p>

<blockquote class="tips">
  <h2 id="numerical-vs-alphanumeric-order">Numerical vs alphanumeric order</h2>
  <p>Sorting by numerical order is useful when dealing with numbers.
In alphanumeric order, 2 comes after 15 because the first character 
“2” comes after first character “1”.
In numerical order, 2 comes before 15, since 2 is smaller than 15.</p>
</blockquote>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">sort</span> <span class="nt">-k2</span> <span class="nt">-n</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008457.1      󠁴̲3󠁴542278 T       G       intergenic_region
CM008465.1      󠁴̲3󠁴999373 G       A       upstream_gene_variant
...
...
CM008455.1      󠁴̲2󠁴89401193       G       A       intergenic_region
CM008455.1      󠁴̲2󠁴94972840       C       T       intergenic_region
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">sort</span> <span class="nt">-k2</span> <span class="nt">-n</span> <span class="nt">-r</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008455.1      294972840       C       T       intergenic_region
CM008455.1      289401193       G       A       intergenic_region
...
...
CM008465.1      3999373 G       A       upstream_gene_variant
CM008457.1      3542278 T       G       intergenic_region
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="argument-shortcuts-in-bash">Argument shortcuts in bash</h2>
  <p>Multiple argument, and sometimes expected values of arguments
can be written together in bash.
In above example, you can replace the above command <code class="language-plaintext highlighter-rouge">-k2 -n -r</code> as <code class="language-plaintext highlighter-rouge">-k2nr</code>. 
Guess what <code class="language-plaintext highlighter-rouge">sort -k2nr file1.txt</code> does.</p>
</blockquote>

<h3 id="replacing-text">Replacing text</h3>

<p><code class="language-plaintext highlighter-rouge">sed</code> command can be used for replacing text in a file. 
<code class="language-plaintext highlighter-rouge">sed 's/old/new/g'</code> replaces all instances of <code class="language-plaintext highlighter-rouge">old</code> with <code class="language-plaintext highlighter-rouge">new</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">sed</span> <span class="s1">'s/CM0084/Chr_/g'</span> tail.txt <span class="o">&gt;</span> prettyfile.txt

<span class="gp">$ </span><span class="nb">cat </span>prettyfile.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Chr_66.1        28255843        A       C       intergenic_region
Chr_65.1        236852266       C       T       intergenic_region
Chr_63.1        16380489        T       A       intergenic_region
Chr_63.1        123496326       T       A       intergenic_region
Chr_57.1        21333612        G       T       intergenic_region
</code></pre></div></div>

<h3 id="search-and-extract">Search and Extract</h3>

<p><code class="language-plaintext highlighter-rouge">grep</code> command is used to find a string in a file and return the matching line.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">grep</span> <span class="s2">"downstream_gene_variant"</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008465.1      57718962        G       A       󠁴̲d󠁴̲o󠁴̲w󠁴̲n󠁴̲s󠁴̲t󠁴̲r󠁴̲e󠁴̲a󠁴̲m󠁴̲_󠁴̲g󠁴̲e󠁴̲n󠁴̲e󠁴̲_󠁴̲v󠁴̲a󠁴̲r󠁴̲i󠁴̲a󠁴̲n󠁴̲t
CM008466.1      236800990       C       A       󠁴̲d󠁴̲o󠁴̲w󠁴̲n󠁴̲s󠁴̲t󠁴̲r󠁴̲e󠁴̲a󠁴̲m󠁴̲_󠁴̲g󠁴̲e󠁴̲n󠁴̲e󠁴̲_󠁴̲v󠁴̲a󠁴̲r󠁴̲i󠁴̲a󠁴̲n󠁴̲t
CM008458.1      226074184       T       TA      󠁴̲d󠁴̲o󠁴̲w󠁴̲n󠁴̲s󠁴̲t󠁴̲r󠁴̲e󠁴̲a󠁴̲m󠁴̲_󠁴̲g󠁴̲e󠁴̲n󠁴̲e󠁴̲_󠁴̲v󠁴̲a󠁴̲r󠁴̲i󠁴̲a󠁴̲n󠁴̲t
CM008458.1      214028749       A       C       󠁴̲d󠁴̲o󠁴̲w󠁴̲n󠁴̲s󠁴̲t󠁴̲r󠁴̲e󠁴̲a󠁴̲m󠁴̲_󠁴̲g󠁴̲e󠁴̲n󠁴̲e󠁴̲_󠁴̲v󠁴̲a󠁴̲r󠁴̲i󠁴̲a󠁴̲n󠁴̲t
CM008460.1      5909607 A       T       󠁴̲d󠁴̲o󠁴̲w󠁴̲n󠁴̲s󠁴̲t󠁴̲r󠁴̲e󠁴̲a󠁴̲m󠁴̲_󠁴̲g󠁴̲e󠁴̲n󠁴̲e󠁴̲_󠁴̲v󠁴̲a󠁴̲r󠁴̲i󠁴̲a󠁴̲n󠁴̲t
CM008460.1      63781473        A       T       󠁴̲d󠁴̲o󠁴̲w󠁴̲n󠁴̲s󠁴̲t󠁴̲r󠁴̲e󠁴̲a󠁴̲m󠁴̲_󠁴̲g󠁴̲e󠁴̲n󠁴̲e󠁴̲_󠁴̲v󠁴̲a󠁴̲r󠁴̲i󠁴̲a󠁴̲n󠁴̲t
CM008465.1      138360632       A       G       󠁴̲d󠁴̲o󠁴̲w󠁴̲n󠁴̲s󠁴̲t󠁴̲r󠁴̲e󠁴̲a󠁴̲m󠁴̲_󠁴̲g󠁴̲e󠁴̲n󠁴̲e󠁴̲_󠁴̲v󠁴̲a󠁴̲r󠁴̲i󠁴̲a󠁴̲n󠁴̲t
</code></pre></div></div>

<p>Argument <code class="language-plaintext highlighter-rouge">-c</code> is used to return the number of matches.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">grep</span> <span class="nt">-c</span> <span class="s2">"downstream_gene_variant"</span> file1.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>7
</code></pre></div></div>

<h3 id="piping-commands">Piping commands</h3>

<p><code class="language-plaintext highlighter-rouge">|</code> operators can be used for piping. 
Piping means that the output of first command serves as input of second command and so on. 
This eliminates need for saving intermediate results as a file.</p>

<p>Lets sort <code class="language-plaintext highlighter-rouge">file1.txt</code> numerically by second column, 
and only display the top 5 lines.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">sort</span> <span class="nt">-k2</span> <span class="nt">-n</span> file1.txt | <span class="nb">head</span> <span class="nt">-n</span> 5
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CM008457.1      3542278 T       G       intergenic_region
CM008465.1      3999373 G       A       upstream_gene_variant
CM008465.1      4979077 A       C       intergenic_region
CM008457.1      5681949 G       A       intergenic_region
CM008460.1      5909607 A       T       downstream_gene_variant
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="exercise-finding-alien-genes-in-the-plant-pathogen-streptomyces-scabies">Exercise: Finding “alien genes” in the plant pathogen <em>Streptomyces scabies</em>.</h2>

  <p>Let’s use the commands in a simple real case. Copy the file <code class="language-plaintext highlighter-rouge">aliens_in_scabies</code> inside <code class="language-plaintext highlighter-rouge">share</code> directory using following command</p>

  <blockquote class="caution">
    <p>Make sure you are in <code class="language-plaintext highlighter-rouge">/blue/general_workshop/&lt;username&gt;</code> directory, 
if not use <code class="language-plaintext highlighter-rouge">cd /blue/general_workshop/&lt;username&gt;</code> to go to your personal working directory.</p>
  </blockquote>

  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cp</span> ../share/strep/aliens_in_scabies ./
</code></pre></div>  </div>

  <p><em>Streptomyces scabies</em> is a plant pathogen that produces necrosis in potatoes. 
Most of the virulence factors are located in regions that have low GC content. 
Also, virulence is highly expressed during the interaction with roots. 
The file “aliens_in_scabies” contains a tabular data with more than 1000 genes 
(first column) along with the level of change expression when growth in rich medium 
vs. interaction with roots (second column). 
GC content% is provided in the third column of the table.</p>

  <p>Using UNIX commands</p>
  <ol>
    <li>Sort the table by expression levels (second column).  <input type="checkbox" /></li>
    <li>Create a new file that contains the top 10 highest expressed genes.  <input type="checkbox" /></li>
    <li>Using the new file, sort the table by gene GC content.  <input type="checkbox" /></li>
    <li>Create a table with only the gene names but replace the name SCAB with SCABIES. <input type="checkbox" /></li>
  </ol>

  <details>
    <summary></summary>

    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># 1</span>
<span class="gp">$ </span><span class="nb">sort</span> <span class="nt">-k2n</span> aliens_in_scabies

<span class="c"># 2</span>
<span class="gp">$ </span><span class="nb">sort</span> <span class="nt">-k2nr</span> aliens_in_scabies | <span class="nb">head</span> <span class="o">&gt;</span> newfile.txt

<span class="c"># 3</span>
<span class="gp">$ </span><span class="nb">sort</span> <span class="nt">-k3n</span> newfile.txt

<span class="c"># 4</span>
<span class="gp">$ </span><span class="nb">cut</span> <span class="nt">-f1</span> aliens_in_scabies | <span class="nb">sed</span> <span class="s1">'s/SCAB/SCABIES/g'</span>
</code></pre></div>    </div>

  </details>
</blockquote>

<h2 id="an-intro-to-loops">An intro to loops</h2>
<p>Many tasks are repetitive. It is not necessary to repeat the same command multiple times. Instead, we can use wildcards and loops to repeat a task.</p>

<h3 id="wildcards">Wildcards</h3>

<p>There are certain symbols in UNIX that represent mutiple values, 
and are called “wildcards”. <code class="language-plaintext highlighter-rouge">*</code> is a universal wildcard that stands for anything.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">ls</span> <span class="k">*</span>.txt
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>concat.txt     files.txt     prettyfile.txt
file1.txt      head.txt      tail.txt
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">mv</span> <span class="k">*</span>.txt demo

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>demo
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">ls </span>demo
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>concat.txt     files.txt     prettyfile.txt
file1.txt      head.txt      tail.txt
</code></pre></div></div>

<p>Other wildcards are</p>

<table>
  <thead>
    <tr>
      <th>Wildcard</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code class="language-plaintext highlighter-rouge">*</code></td>
      <td>any character or characters</td>
    </tr>
    <tr>
      <td><code class="language-plaintext highlighter-rouge">?</code></td>
      <td>any single character</td>
    </tr>
    <tr>
      <td><code class="language-plaintext highlighter-rouge">[0-9]</code></td>
      <td>any single number</td>
    </tr>
    <tr>
      <td><code class="language-plaintext highlighter-rouge">[A-Z]</code></td>
      <td>any single capital alphabets</td>
    </tr>
    <tr>
      <td><code class="language-plaintext highlighter-rouge">[a-z]</code></td>
      <td>aany single small alphabets</td>
    </tr>
    <tr>
      <td><code class="language-plaintext highlighter-rouge">[!x]</code></td>
      <td>not x</td>
    </tr>
    <tr>
      <td><code class="language-plaintext highlighter-rouge">{x, y}</code></td>
      <td>x or y</td>
    </tr>
  </tbody>
</table>

<h3 id="the-for-loop">The ‘FOR’ loop</h3>

<p><code class="language-plaintext highlighter-rouge">For</code> loop iterates over a specific range or numbers or an array. For example.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="k">for </span>x <span class="k">in</span> <span class="o">{</span>0..3<span class="o">}</span>
<span class="gp">&gt; </span><span class="k">do</span>
<span class="gp">&gt; </span>  <span class="nb">echo</span> <span class="nv">$x</span>
<span class="gp">&gt; </span><span class="k">done</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>0
1
2
3
</code></pre></div></div>

<p>The process between <code class="language-plaintext highlighter-rouge">do</code> and <code class="language-plaintext highlighter-rouge">done</code> enclose task to be repeated.
The variable <code class="language-plaintext highlighter-rouge">x</code> will take consecutive values from array <code class="language-plaintext highlighter-rouge">{0..3}</code> in each iteration.</p>

<blockquote class="tips">
  <p>Try <code class="language-plaintext highlighter-rouge">echo {0..3}</code> to see what <code class="language-plaintext highlighter-rouge">{0..3}</code> represents.</p>
</blockquote>

<blockquote class="tips">
  <h2 id="bash-prompts">Bash prompts</h2>
  <p><code class="language-plaintext highlighter-rouge">$</code> prompt specifies ready to start a command.
<code class="language-plaintext highlighter-rouge">#</code> replaces <code class="language-plaintext highlighter-rouge">$</code> for root user.
<code class="language-plaintext highlighter-rouge">&gt;</code> specifies continuation of multiline command from previous line.</p>
</blockquote>

<h3 id="the-while-loop">The ‘WHILE’ loop</h3>

<p><code class="language-plaintext highlighter-rouge">while</code> loop iterates as long as a condition is true. Lets use the example.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nv">x</span><span class="o">=</span>0
<span class="gp">$ </span><span class="k">while</span><span class="o">((</span><span class="nv">$x</span> &lt; 4<span class="o">))</span>
<span class="gp">&gt; </span><span class="k">do</span> 
<span class="gp">&gt; </span>  <span class="nb">echo</span> <span class="nv">$x</span>
<span class="gp">&gt; </span>  <span class="o">((</span>++x<span class="o">))</span> 
<span class="gp">&gt; </span><span class="k">done</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>0
1
2
3
</code></pre></div></div>

<p>As in <code class="language-plaintext highlighter-rouge">for</code> loop, the commands to be looped over are enclosed between <code class="language-plaintext highlighter-rouge">do</code> and <code class="language-plaintext highlighter-rouge">done</code>.</p>

<blockquote class="notes">
  <h2 id="learn-more-about-unix-commands">Learn more about unix commands</h2>
  <p><a href="https://swcarpentry.github.io/shell-novice/reference/">Software carpentary reference</a></p>
</blockquote>





<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>pwd, ls, and cd are three main commands that allow navigation through directories</p>
</li>
    
    <li><p>cat, more and less display text files</p>
</li>
    
    <li><p>sed and grep are command for replacing and extracting characters in a text file.</p>
</li>
    
    <li><p>cut and sort are commands to edit text files.</p>
</li>
    
    <li><p>Unix commands allow handling repetitive tasks, suing for or while loops.</p>
</li>
    
  </ul>
</blockquote>

</article>



























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../"><span class="glyphicon glyphicon-menu-up" aria-hidden="true"></span><span class="sr-only">lesson home</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../02-slurm/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>


      
      
<footer>
  <hr/>
  <div class="row">
    <div class="col-md-6" align="left">
      <h4>
	      Copyright &copy; 2020–2022 <a href="https://github.com/BioInfoAPS">APS Basic Bioinformatics Workshop Team</a>
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>

<script>
/* Based on https://stackoverflow.com/a/30894829/1802726 */
var show_instructions = function() {
     //set all nav tabs to inactive
    $('.nav-tabs li').removeClass('active');

    //get all nav tabs matching the os and set to active
    $('.nav-tabs li a[href*="'+localStorage.os+'"]').closest('li').addClass('active');

    //set all tabs to inactive
    $('.tab-pane').removeClass('active');

    //get all tabs matching the os and set to active
    $('.tab-pane[id*="'+localStorage.os+'"]').addClass('active');
}

if (typeof(Storage) !== "undefined") {
    if (localStorage.os == "undefined") {
        localStorage.os = "windows";
    }
    show_instructions();
} else {
    console.log("No localStorage");
}

/* Based on https://stackoverflow.com/a/30894829/1802726 */
$('.nav-tabs li a').click(function(){
    //get selected os
    localStorage.os = $(this)[0].dataset.os;
    show_instructions();
});

  // detect misssing iframe for Eventbrite
  $(document).ready(function(){
      if ($('iframe').height() == 0) {
          $('iframe').before('<p>Looks like your adblocker blocked the registration window. Please navigate to <a href="https://www.eventbrite.com/tickets-external?eid=&ref=etckt">https://www.eventbrite.com/tickets-external?eid=&ref=etckt</a> to register.</p>');
      }
  });

  // hide alerts when site is rendered on carpentries.github.io
  $(document).ready(function() {
      if (location.href.startsWith("https://carpentries.github.io/workshop-template/")) {
          $("div.alert").hide();
      }
  });

</script>

  </body>
</html>
