















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-09-22 19:36:24 -0400">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="https://bioinfoaps.github.io/">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    <link rel="license" href="#license-info" />

    

    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Basic	bioinformatics and command-line	tools	for phytopathologists"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  BLAST search &ndash; Basic	bioinformatics and command-line	tools	for phytopathologists
  </title>  

  </head>
  <body>

    <div class="container">
      
















  
  















<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      <a href="" class="pull-left">
        <img class="navbar-logo" src="../assets/logo/icon.svg" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	      
        <li><a href="../setup.html">Setup</a></li>

        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Lessons <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-unix/index.html">UNIX command line</a></li>
            
            
            <li><a href="../02-slurm/index.html">SLURM workload manager</a></li>
            
            
            <li><a href="../03-blast/index.html">BLAST search</a></li>
            
            
            <li><a href="../04-parse/index.html">Parsing BLAST output</a></li>
            
            
            <li><a href="../05-tree/index.html">Phylogenetic tree</a></li>
            
          </ul>
        </li>

        

      

      <li class="dropdown">
        <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Optional <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../10-more_unix/index.html">More UNIX commands</a></li>
          
            <li><a href="../20-text_process/index.html">FASTA manipulation with awk</a></li>
          
            <li><a href="../30-extra_tools/index.html">Another bioinformatic tool: PyANI</a></li>
          
            <li><a href="../40-awesome_bioinfo/index.html">Awesome bioinformatics</a></li>
          
            <li><a href="../50-multiple_gene_copies/index.html">Multiple gene copies</a></li>
          
        </ul>
      </li>

      

      <li class="dropdown">
      <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../about/">Authors</a></li>
          
            <li><a href="../acknowledgments/">Acknowledgements</a></li>
          
            <li><a href="../faq/index.html">FAQ</a></li>
          
            <li><a href="../intro_slides/index.html">Intro Slides</a></li>
          
        </ul>
      </li>

      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" title="Google search" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>




























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../02-slurm/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Basic	bioinformatics and command-line	tools	for phytopathologists</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../04-parse/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">BLAST search</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>












<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 25 min
      <br/>
      <strong>Exercises:</strong> 35 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can we conduct local blast searches using the terminal?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Learn how to conduct blast searches locally using the terminal</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p><strong>B</strong>asic <strong>L</strong>ocal <strong>A</strong>lignment <strong>S</strong>earch <strong>T</strong>ool (<a href="https://doi.org/10.1016/s0022-2836(05)80360-2" target="_blank">Altschul S et al.</a>) is a set of programs 
that search sequence database for statistically significant similarities. 
There are five traditional BLAST programs available in BLAST+ tools in commandline:</p>

<table>
  <thead>
    <tr>
      <th>Program</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>BLASTN</td>
      <td>compares a nucleotide query sequence against a nucleotide sequence database.</td>
    </tr>
    <tr>
      <td>BLASTP</td>
      <td>compares an amino acid query sequence against a protein sequence database.</td>
    </tr>
    <tr>
      <td>BLASTX</td>
      <td>compares a nucleotide query sequence translated in a ll reading frames against a protein sequence database.</td>
    </tr>
    <tr>
      <td>TBLASTN</td>
      <td>compares a protein query sequence against a nucleotide sequence database dynamically translated in all reading frames.</td>
    </tr>
    <tr>
      <td>TBLASTX</td>
      <td>compares the six-frame translations of a nucleotide query sequence against the six-frame translations of a nucleotide sequence database.</td>
    </tr>
  </tbody>
</table>

<p>These are the same blast applications what you can see in the <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi" target="_blank">NCBI blast website</a>. 
Read more about command line <a href="https://open.oregonstate.education/computationalbiology/chapter/command-line-blast/" target="_blank">BLAST 
here</a>.</p>

<h2 id="blast-terminology">BLAST terminology</h2>

<ul>
  <li>Query
    <ul>
      <li>The sequence for which the search is performed.</li>
    </ul>
  </li>
  <li>Database
    <ul>
      <li>A set of sequences on which search is performed to find matches for the query sequence.</li>
    </ul>
  </li>
  <li>Subject
    <ul>
      <li>A single sequence within the database.</li>
    </ul>
  </li>
  <li>Score (S)
    <ul>
      <li>A value representing quality of each alignment.</li>
      <li>It is calculated as the sum of substitution and gap scores.</li>
    </ul>
  </li>
  <li>Expectation value (E-value or E).
    <ul>
      <li>A value representing probabilistic significance of each alignment.</li>
      <li>A number of different alignments with scores equivalent to or better than ‘S’ that are expected to occur in database search by chance.</li>
      <li>The lower the E value, the more significant the score.</li>
    </ul>
  </li>
</ul>

<h2 id="application-of-blast">Application of BLAST</h2>

<p>BLAST can be run online (e.g. NCBI blast) as well as locally. Few applications of local BLAST are:</p>

<ol>
  <li>to compare against multiple genomes of interest,</li>
  <li>to output in different formats that can be used in downstream applications.</li>
</ol>

<h2 id="outline-of-sequences-to-blast">Outline of sequences to BLAST</h2>

<p>We will use the local BLAST to identify if the bacterial genomes carry a gene of interest. 
We will use bacterial spot causing <em>Xanthomonas</em> and one of the conserved effector ‘AvrBs2’ sequence as an example today. More information on <em>Xanthomonas</em> is <a href="https://www.nature.com/articles/s41579-020-0361-8">available here</a>.</p>

<blockquote class="notes">
  <h2 id="pathogen-effectors-proteins">Pathogen effectors proteins</h2>
  <p>Effectors are proteins that are secreted by 
pathogenic bacteria into the plant cells to. 
manipulate host prcoesses and achieve colonization. 
The effector gene we are analyzing in this course 
named <a href="https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1002408"><em>avrBs2</em></a> is one of the most studied bacterial effectors 
because of its significant virulence contribution and is 
conserved across multiple <em>Xanthomonas</em> species.</p>
</blockquote>

<p>In the folder <code class="language-plaintext highlighter-rouge">/blue/share/xanthomonas/</code>, there are four <em>Xanthomonas</em> species genomes.</p>

<ul>
  <li><em>Xanthomonas euvesicatoria</em> → <code class="language-plaintext highlighter-rouge">Xeu.fasta</code></li>
  <li><em>Xanthomonas perforans</em> → <code class="language-plaintext highlighter-rouge">Xp.fasta</code></li>
  <li><em>Xanthomonas gardneri</em> → <code class="language-plaintext highlighter-rouge">Xg.fasta</code></li>
  <li><em>Xanthomonas citri</em> → <code class="language-plaintext highlighter-rouge">Xc.fasta</code></li>
</ul>

<p>The folder also contains nucleotide sequence for <em>avrBs2</em> gene as <code class="language-plaintext highlighter-rouge">avrBs2.fas</code>.</p>

<p>Our objective would be to see if the bacterial genomes carry the effector gene.</p>

<p>Before starting, copy all the required files to your working directory.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cd</span> /blue/general_workshop/&lt;username&gt;

<span class="gp">$ </span><span class="nb">cp</span> <span class="nt">-r</span> ../share/xanthomonas ./

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>demo     strep     slurm     xanthomonas
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cd </span>xanthomonas

<span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>avrBs2.fas     Xc.fasta     Xeu.fasta     Xg.fasta     Xp.fasta
</code></pre></div></div>

<h3 id="loading-blast-in-hipergator-cluster">Loading BLAST in Hipergator cluster</h3>

<p>First, we need the BLAST program. 
BLAST tools is already installed in HiperGator, but not available by default.
We can use <code class="language-plaintext highlighter-rouge">module load</code> command to load the program.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>module load ncbi_blast
</code></pre></div></div>

<h3 id="creating-blast-database">Creating BLAST database</h3>

<p>Next, we need to specify the database to blast against, 
which in our case are the genome files <code class="language-plaintext highlighter-rouge">Xeu.fasta</code>, <code class="language-plaintext highlighter-rouge">Xp.fasta</code>, <code class="language-plaintext highlighter-rouge">Xg.fasta</code>, and <code class="language-plaintext highlighter-rouge">Xc.fasta</code>.
<code class="language-plaintext highlighter-rouge">makeblastdb</code> command, which is a part of BLAST+ package, 
is used for creating BLAST databases from FASTA files.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>makeblastdb <span class="nt">-in</span> Xeu.fasta <span class="nt">-out</span> Xeu <span class="nt">-dbtype</span> nucl
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Building a new DB, current time: 09/15/2020 02:25:22
New DB name:   /blue/general_workshop/&lt;username&gt;/xanthomonas/Xeu
New DB title:  Xeu.fasta
Sequence type: Nucleotide
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 3 sequences in 0.0926349 seconds.
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">ls</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>avrBs2.fas     Xeu.ndb     Xeu.not     Xeu.ntf     Xg.fasta
Xc.fasta       Xeu.nhr     Xeu.nsq     Xeu.nto     Xp.fasta
Xeu.fasta      Xeu.nin
</code></pre></div></div>

<blockquote class="notes">
  <p>New database files with extensions <code class="language-plaintext highlighter-rouge">.ndb</code>, <code class="language-plaintext highlighter-rouge">.nhr</code>, <code class="language-plaintext highlighter-rouge">.nsq</code> etc. have been created.</p>
</blockquote>

<blockquote class="tips">
  <h2 id="understanding-command-arguments">Understanding command arguments</h2>
  <p>For most unix commandline commands, you can use <code class="language-plaintext highlighter-rouge">-h</code> or <code class="language-plaintext highlighter-rouge">--help</code> argument/flag to see help or options.
This will include list of arguments and their meanings.
In this case, <code class="language-plaintext highlighter-rouge">makeblastdb -h</code> command displays options avaiable for <code class="language-plaintext highlighter-rouge">makeblastdb</code>.<br />
Alternatively, <code class="language-plaintext highlighter-rouge">man &lt;command&gt;</code> shows manual for the command.</p>
</blockquote>

<h3 id="creating-blast-database-in-batch">Creating BLAST database in batch</h3>

<p>Making database individually for all genomes will be manually laborious. 
Instead, we can identify common patterns in the name of the genome files and 
execute <code class="language-plaintext highlighter-rouge">makeblastdb</code> in a loop.</p>

<blockquote class="challenge">
  <p>What is one pattern that is common in all the genome files?</p>
  <details>
<summary></summary>
The extension - ‘.fasta’.
</details>
</blockquote>

<p>We can specify all files in current path with fasta extension by <code class="language-plaintext highlighter-rouge">./*.fasta</code>.</p>

<blockquote class="caution">
  <p>The command below has backticks <code class="language-plaintext highlighter-rouge">`</code>, 
which is the symbol just left to <kbd>1</kbd> key and 
just above the <kbd>Tab</kbd> key in the keyboard. 
Do not mistake it for quotes.</p>
</blockquote>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nv">genomes</span><span class="o">=</span><span class="sb">`</span><span class="nb">ls</span> <span class="k">*</span>.fasta | <span class="nb">sed</span> <span class="s1">'s/.fasta//g'</span><span class="sb">`</span>

<span class="gp">$ </span><span class="nb">echo</span> <span class="nv">$genomes</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Xc Xeu Xg Xp
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="k">for </span>genome <span class="k">in</span> <span class="nv">$genomes</span><span class="p">;</span> <span class="k">do </span>makeblastdb <span class="nt">-in</span> <span class="s2">"</span><span class="nv">$genome</span><span class="s2">.fasta"</span> <span class="nt">-out</span> <span class="nv">$genome</span> <span class="nt">-dbtype</span> nucl<span class="p">;</span> <span class="k">done</span>
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>...
...
</code></pre></div></div>

<blockquote class="tips">
  <h2 id="enclosing-commands-in--backticks">Enclosing commands in <code class="language-plaintext highlighter-rouge">`</code> (backticks)</h2>
  <p><code class="language-plaintext highlighter-rouge">`</code> act like brackets in math.
First, the commands inside backticks are executed, 
and then the output is used for commands outside.</p>
</blockquote>

<blockquote class="tips">
  <h2 id="one-liners">One liners</h2>
  <p>Short loops can be written in a same line by separating commands with <code class="language-plaintext highlighter-rouge">;</code>. 
<code class="language-plaintext highlighter-rouge">;</code> is equivalent to pressing <kbd>Enter</kbd>.</p>
</blockquote>

<blockquote class="notes">
  <p><code class="language-plaintext highlighter-rouge">sed</code> command is being used to remove <code class="language-plaintext highlighter-rouge">.fasta</code> extension from list of names.</p>
</blockquote>

<blockquote class="tips">
  <p>For maually selecting the databases, you can use <code class="language-plaintext highlighter-rouge">for</code> loop like this:
<code class="language-plaintext highlighter-rouge">for genome in Xp Xg Xc; do makeblastdb -in "$genome.fasta" -out $genome -dbtype nucl; done</code></p>
</blockquote>

<blockquote class="notes">
  <h2 id="merging-database-with-blastdb_aliastool">Merging database with <code class="language-plaintext highlighter-rouge">blastdb_aliastool</code></h2>
  <p>BLAST+ also includes a command <code class="language-plaintext highlighter-rouge">blastdb_aliastool</code> for combining databases; 
however, it is outside the scope of this workshop.<br />
Usage:<br />
<code class="language-plaintext highlighter-rouge">blastdb_aliastool -dblist "Xeu Xp Xg Xc" -dbtype nucl -out Xspp -title "Xanthomonas genomic"</code></p>
</blockquote>

<h3 id="performing-blast-search">Performing BLAST search</h3>

<p>We are now ready to do a BLAST search. Since both the query (<code class="language-plaintext highlighter-rouge">avrBs2.fasta</code>) 
and the database are nucleotide sequences, we will perform <code class="language-plaintext highlighter-rouge">blastn</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>blastn <span class="nt">-query</span> avrBs2.fas <span class="nt">-db</span> Xeu <span class="nt">-out</span> Xeu_avrBs2.out <span class="nt">-evalue</span> 0.001

<span class="gp">$ </span><span class="nb">ls</span> <span class="k">*</span>.out
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Xeu_avrBs2.out
</code></pre></div></div>

<h3 id="performing-blast-search-in-multiple-databases-in-batch">Performing BLAST search in multiple databases in batch</h3>

<p>We can blast multiple databases in a loop as well. 
Let’s do this in a different way from previous example.
First lets create a list of all databases to BLAST against and save it into a file.
We can use <code class="language-plaintext highlighter-rouge">nano</code> for writing to a file.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span>nano
</code></pre></div></div>

<p>A basic text editor will open. Type in all the databases to search against.</p>
<div class="language-plaintext terminal highlighter-rouge"><div class="highlight"><pre class="highlight"><code>-----------------------------------------------------------------------------------------------
 GNU nano 3.3 beta 02                      New Buffer
-----------------------------------------------------------------------------------------------
Xp
Xg
Xc

-----------------------------------------------------------------------------------------------
^G Get Help     ^O WriteOut     ^R Read File     ^Y Prev Page     ^K Cut Text       ^C Cur Pos
^X Exit         ^J Justify      ^W Where Is      ^V Next Page     ^U UnCut Text     ^T To Spell
-----------------------------------------------------------------------------------------------
</code></pre></div></div>

<p>You do not have to press enter and go to the next line after typing ‘Xc’. The next line will be read as an empty line and give error message. Although the additional line and error message due to the empty line does not affect our outputs here, it might be relevant for other situations.</p>

<p>Press <kbd>Ctrl</kbd>+<kbd>o</kbd> to save the file.
Give it a name <code class="language-plaintext highlighter-rouge">dblist.txt</code> and press <kbd>Enter</kbd>.</p>

<p>Press <kbd>Ctrl</kbd>+<kbd>x</kbd> to return to bash prompt.</p>

<p>Now we can run the <code class="language-plaintext highlighter-rouge">blastn</code> in loop.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="k">for </span>dbname <span class="k">in</span> <span class="sb">`</span><span class="nb">cat </span>dblist.txt<span class="sb">`</span>
<span class="gp">&gt; </span><span class="k">do</span>
<span class="gp">&gt; </span>  blastn <span class="nt">-query</span> avrBs2.fas <span class="nt">-db</span> <span class="s2">"</span><span class="nv">$dbname</span><span class="s2">"</span> <span class="nt">-out</span> <span class="nv">$dbname</span><span class="s2">"_avrBs2.out"</span> <span class="nt">-evalue</span> 0.001
<span class="gp">&gt; </span><span class="k">done</span> &lt; dblist.txt
</code></pre></div></div>

<blockquote class="caution">
  <p>After copying the code block, do not forget to remove the <code class="language-plaintext highlighter-rouge">$</code> or <code class="language-plaintext highlighter-rouge">&gt;</code> sign.
Or click <a href="/code/blastn_loop.txt" target="_blank">here</a> to open a separate page to copy the code.</p>
</blockquote>

<blockquote class="challenge">
  <h2 id="exercise-performing-blast-search-in-slurm">Exercise: Performing blast search in SLURM</h2>

  <p>We can also run <code class="language-plaintext highlighter-rouge">blast</code> as a SLURM job, which is useful for long and resource intensive search.</p>

  <p>The SLURM submission has been prepared for your and is available as 
<code class="language-plaintext highlighter-rouge">/blue/general_workshop/share/scripts/slurm_blast.sh</code>. 
Genome and query files are available in <code class="language-plaintext highlighter-rouge">/blue/general_workshop/share/xanthomonas</code>.
You can use the checklist to mark progress.</p>

  <ol>
    <li>Change your location to your working directory <code class="language-plaintext highlighter-rouge">/blue/general_workshop/&lt;username&gt;</code>. <input type="checkbox" /></li>
    <li>Make a folder in your working directory called <code class="language-plaintext highlighter-rouge">slurm_blast</code> and enter that directory. <input type="checkbox" /></li>
    <li>Copy all genome and query fasta files to current directory. <input type="checkbox" /></li>
    <li>Copy the submission script to the current directory. <input type="checkbox" /></li>
    <li>Open the script in nano and edit the email address. <input type="checkbox" /></li>
    <li>Submit the job to SLURM. <input type="checkbox" /></li>
    <li>After job is completed, check if output files exist in current directory. <input type="checkbox" /></li>
  </ol>

  <details>
    <summary></summary>

    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#1</span>
<span class="gp">$ </span><span class="nb">cd</span> /blue/general_workshop/&lt;username&gt;

<span class="c">#2</span>
<span class="gp">$ </span><span class="nb">mkdir </span>slurm_blast

<span class="gp">$ </span><span class="nb">cd </span>slurm_blast

<span class="c">#3</span>
<span class="gp">$ </span><span class="nb">cp</span> ../../share/xanthomonas/<span class="k">*</span> ./

<span class="c">#4</span>
<span class="gp">$ </span><span class="nb">cp</span> ../../share/scripts/slurm_blast.sh ./

<span class="c">#5</span>
<span class="gp">$ </span>nano slurm_blast.sh
→ edit email address → ctrl+x → y

<span class="c">#6</span>
<span class="gp">$ </span>sbatch slurm_blast.sh

<span class="c">#7</span>
<span class="gp">$ </span><span class="nb">ls</span> <span class="k">*</span>.out
</code></pre></div>    </div>

  </details>
</blockquote>





<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Using UNIX commands, we can customize blast searches.</p>
</li>
    
  </ul>
</blockquote>

</article>



























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../02-slurm/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../04-parse/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>


      
      
<footer>
  <hr/>
  <div class="row">
    <div class="col-md-6" align="left">
      <h4>
	      Copyright &copy; 2020–2022 <a href="https://github.com/BioInfoAPS">APS Basic Bioinformatics Workshop Team</a>
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>

<script>
/* Based on https://stackoverflow.com/a/30894829/1802726 */
var show_instructions = function() {
     //set all nav tabs to inactive
    $('.nav-tabs li').removeClass('active');

    //get all nav tabs matching the os and set to active
    $('.nav-tabs li a[href*="'+localStorage.os+'"]').closest('li').addClass('active');

    //set all tabs to inactive
    $('.tab-pane').removeClass('active');

    //get all tabs matching the os and set to active
    $('.tab-pane[id*="'+localStorage.os+'"]').addClass('active');
}

if (typeof(Storage) !== "undefined") {
    if (localStorage.os == "undefined") {
        localStorage.os = "windows";
    }
    show_instructions();
} else {
    console.log("No localStorage");
}

/* Based on https://stackoverflow.com/a/30894829/1802726 */
$('.nav-tabs li a').click(function(){
    //get selected os
    localStorage.os = $(this)[0].dataset.os;
    show_instructions();
});

  // detect misssing iframe for Eventbrite
  $(document).ready(function(){
      if ($('iframe').height() == 0) {
          $('iframe').before('<p>Looks like your adblocker blocked the registration window. Please navigate to <a href="https://www.eventbrite.com/tickets-external?eid=&ref=etckt">https://www.eventbrite.com/tickets-external?eid=&ref=etckt</a> to register.</p>');
      }
  });

  // hide alerts when site is rendered on carpentries.github.io
  $(document).ready(function() {
      if (location.href.startsWith("https://carpentries.github.io/workshop-template/")) {
          $("div.alert").hide();
      }
  });

</script>

  </body>
</html>
