















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-10-17 15:07:32 -0400">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="https://bioinfoaps.github.io/">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    <link rel="license" href="#license-info" />

    

    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Basic	bioinformatics and command-line	tools	for phytopathologists"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  FASTA manipulation with awk &ndash; Basic	bioinformatics and command-line	tools	for phytopathologists
  </title>  

  </head>
  <body>

    <div class="container">
      
















  
  















<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      <a href="" class="pull-left">
        <img class="navbar-logo" src="../assets/logo/icon.svg" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	      
        <li><a href="../setup.html">Setup</a></li>

        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Lessons <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-unix/index.html">UNIX command line</a></li>
            
            
            <li><a href="../02-slurm/index.html">SLURM workload manager</a></li>
            
            
            <li><a href="../03-blast/index.html">BLAST search</a></li>
            
            
            <li><a href="../04-parse/index.html">Parsing BLAST output</a></li>
            
            
            <li><a href="../05-tree/index.html">Phylogenetic tree</a></li>
            
          </ul>
        </li>

        

      

      <li class="dropdown">
        <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Optional <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../10-more_unix/index.html">More UNIX commands</a></li>
          
            <li><a href="../20-text_process/index.html">FASTA manipulation with awk</a></li>
          
            <li><a href="../30-extra_tools/index.html">Another bioinformatic tool: PyANI</a></li>
          
            <li><a href="../40-awesome_bioinfo/index.html">Awesome bioinformatics</a></li>
          
            <li><a href="../50-multiple_gene_copies/index.html">Multiple gene copies</a></li>
          
        </ul>
      </li>

      

      <li class="dropdown">
      <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
        <ul class="dropdown-menu">
          
            <li><a href="../about/">About</a></li>
          
            <li><a href="../acknowledgments/">Acknowledgements</a></li>
          
            <li><a href="../faq/index.html">FAQ</a></li>
          
            <li><a href="../intro_slides/index.html">Intro Slides</a></li>
          
        </ul>
      </li>

      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" title="Google search" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>




























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../10-more_unix/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Basic	bioinformatics and command-line	tools	for phytopathologists</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../30-extra_tools/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">FASTA manipulation with awk</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>






<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>To process FASTA records, specifically FASTA filtration using awk.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h2 id="the-fasta-format">The FASTA format</h2>

<p>One of the most common file format when working in bioinformatics is
the FASTA file. FASTA format holds a nucleotide or amino acid sequences,
following a (unique) identifier, called a description line.
A FASTA file is a text file, often with extension <code class="language-plaintext highlighter-rouge">.fasta</code>, <code class="language-plaintext highlighter-rouge">.fas</code>, or <code class="language-plaintext highlighter-rouge">.fa</code>
for both nucleotide and proteins, <code class="language-plaintext highlighter-rouge">.fna</code> for nucleotide only,
<code class="language-plaintext highlighter-rouge">.faa</code> for amino acid sequence only.
A multiFASTA file is the concatenation of multiple
FASTA sequences of one type (nucleotide or amino acid).</p>

<p>The description line precedes the sequence.
It starts with a <code class="language-plaintext highlighter-rouge">&gt;</code> followed by the sequence identifier and
is delimited within a single line for each sequence.
<code class="language-plaintext highlighter-rouge">&gt;</code> sign, in general, does not appear anywhere else
except in the beginning of the description line.
The sequence starts from immediately next line, and long sequences are
often into many lines often with ~80 characters/bases per line.
Both uppercase and lowercase characters are acceptable,
although some programs assign meaning to the case.
Degenerate bases are accepted, so are <code class="language-plaintext highlighter-rouge">-</code> for gaps and <code class="language-plaintext highlighter-rouge">*</code> for stop codon.
Any other characters, including spaces, are ignored.</p>

<p>Here is an example of a valid FASTA file.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">cat </span>files/fasta/example.fasta
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;Xanthomonas perforans strain GV872 contig_50
TGCGGGCGGTGAGGTCGCCCGCCGCGATCGCCTGCAATACCGTCGAGAGCGCGTGCAGATTCGCATCGGAGGTGGCCATC
AATTGATTGAGGCTGTCCACCATCGCGCGGAAGTCGTACTGGAAGCGCTGTGCGTCGCCGCGCACGCTGAAATCGCCATT
GGCTGCGGCCTGCGCAAGCAGCTTGATCTCGCTGTTCATCGCCGACAGGTTG
&gt;Xanthomonas perforans strain GV872 contig_51
CCCTCCACAGGAAGCGGACCGGACAATGACGCTTGATTGCCGCGCAGGAAATGCCCCCGTTCCCGCAGGCATACACCCCA
TTTGCCGCGGACATACGCCGCACATTTCTCGGAGCCTGGACAAGTCGCCACAGCTTTGATGGAACGACGGAAAAAACGCC
CCGGAGTTGACCAGAACAACCGCCAGTACTGAAATCTGGACCTTAAGTTCCAGTTCCTGCAGCTCTGGACCTGGTTCATC
AGCGAGCTTCTATCGCTTCGCGGCGGCTTCGATCGCAGCGTGTGCGCGCCATCACTTGGTCGTGCGAAATCGTCGGCCGC
GTGTCGGCCACGGCGCGCTCGATCTTCGCCCGGAACTACCAGTCGTAGTCCTCGGCGGCTTCGGTCGACTCAAATTCGGA
AACAGGTGGTCGAGTGATACGCCCAGGGTTC
&gt;Xanthomonas perforans strain GV872 contig_52
CGCCAGCGCAGCCGACCACGACGCCCGAATCCAGCGTGCCGGCCTCGTTGCCGCCCGCGTCCACGTCGGGCAATTAAGCC
GCGCCCCGCGTTTCCGGTCATGGAACTTCGCCCAGAGCAGCTGGCCGGCCAGTCCAGCCAGCCATTGCAGCCGGTGTATC
TGATCGCCGGCCCCGAGACCTTGCGTGTGCTTGAGGCCGCCGATGCGGTGCGCGCGCGCGCGCGCGC
&gt;Xanthomonas perforans strain GV872 plasmid_contig_53
GGCGTTGTTCCAGCCCAGCACCATTTCCAACAGATTCAGGGCCTGCGCCGGCGACACCGGATTGGGCGTGGCCAAGGTCA
CGGTGCCCTGCACGCCCGGCGCGATGACGTAGTTCTGGCCGAGCATGTCGCCCAGGATCGCCTTGACCACCGCCTGCAGC
GATTCGCCTTCGAAATTGAAGGTGGCGCTGCCGCTGCTGGCCATGCCCAGCGTG
&gt;Xanthomonas perforans strain GV872 plasmid_contig_54
AAATCCTGGGCGTATCAGACCGCTGCCGAGCGTTTGCTTTGGTACCGCCTGCGTGATCGGCGTCCCCCC/àùbä';ç3cC
¦Å¼OÈ’LäÖ’sëmôzÒÒ${–-ÃêFY;+ã}:ãž0}%ÂZDÅì9D‹Rž÷D×ç°:âñÆó²9h‰ÐžæK¼ë™XÕJK¥·—(ü7¢]ÃÆ
Ño5®o&amp;ÇDçJ^ÃŒZòsÀš‹Ü´†ü¿ÄŒ—xÎ—¸¬µ»?à=Of±)2?îP3ñLapÈëY_{DŸm)ã}ž[Êº„ódMÅbóSl¤ÌŽÑ‚¼
Óh+¶\Éz½/­Z¹â²kgfSû%äÖÿzvŽ#?69éhÎÔøK×‰Êí¨/4ŠSVfÍõÔûÁÒ&amp;½ûE÷kÑ²:k ÷yxÙgH2/ÂžQ,[@ûIß
€|Ë²Ï2,ôCÆ
</code></pre></div></div>

<h3 id="fasta-file-processing">FASTA file processing</h3>

<p>The simplicity of FASTA file makes it a good target for learning text processing
for beginner bioinformaticians. To be clear, there are plethora of tools and 
programs that can be used to achieve all kinds of processing
(such as filtration, editing, metadata computation, etc.).
Rather, the objective of this section is to learn to do those processing 
by writing your own code.</p>

<p>There are many tools which can be achieve same processing steps.
In this section, we will use a powerful text processing utility called <code class="language-plaintext highlighter-rouge">awk</code>, 
which is available by default in all modern UNIX systems.</p>

<h2 id="processing-a-file-in-awk">Processing a file in <code class="language-plaintext highlighter-rouge">awk</code></h2>

<p>The syntax for awk to process a file looks like follows.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">awk</span> <span class="s1">'
    commands to process ...
    ...
  '</span> input_file

<span class="gp">$ </span><span class="nb">cat </span>input_file | <span class="nb">awk</span> <span class="s1">'
    commands to process ...
    ...
  '</span>
</code></pre></div></div>

<h3 id="processing-one-fasta-sequence-at-a-time">Processing one FASTA sequence at a time</h3>

<p>Despite its simple structure, processing a FASTA using <code class="language-plaintext highlighter-rouge">awk</code>
(or most other UNIX tools) poses one important challenge.
<code class="language-plaintext highlighter-rouge">Awk</code> processes one line at a time, but ideally,
we’d like to process one FASTA sequence at a time.
Fortunately, there are ways to handle this in <code class="language-plaintext highlighter-rouge">awk</code>.</p>

<p>Consider a simple table:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>abc def ghi
jkl mno pqr
stu vwx yz
</code></pre></div></div>

<p>In awk, columns are called fields and rows are called records.
In the table above, space character ` ` separates different columns (or fields)
and is called a field separator.
Similarly, newline character, represented by <code class="language-plaintext highlighter-rouge">\n</code>, separates different rows
(or records) and is called a record separator.
Input and output files can have different field and record separators in <code class="language-plaintext highlighter-rouge">awk</code>.
They are stored in following variables within <code class="language-plaintext highlighter-rouge">awk</code>:</p>

<table>
  <thead>
    <tr>
      <th>Variables</th>
      <th>Meaning</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>FS</td>
      <td>Field separator (input, and output too if OFS not specified)</td>
    </tr>
    <tr>
      <td>OFS</td>
      <td>Output field separator</td>
    </tr>
    <tr>
      <td>RS</td>
      <td>Record separator (input, and output too if ORS not specified)</td>
    </tr>
    <tr>
      <td>ORS</td>
      <td>Output record separator</td>
    </tr>
  </tbody>
</table>

<p>A individual FASTA record starts with a <code class="language-plaintext highlighter-rouge">&gt;</code> character, so if we use <code class="language-plaintext highlighter-rouge">&gt;</code> as 
the record separator, then <code class="language-plaintext highlighter-rouge">awk</code> would process one entire FASTA at a time.</p>

<blockquote class="notes">
  <h2 id="accessing-fields-in-awk">Accessing fields in <code class="language-plaintext highlighter-rouge">awk</code></h2>

  <p>Field values can be accessed in <code class="language-plaintext highlighter-rouge">awk</code> using numbered variables.
For the record currently being processed,</p>
  <ul>
    <li><code class="language-plaintext highlighter-rouge">$0</code> stores the entire record</li>
    <li><code class="language-plaintext highlighter-rouge">$1</code> stores first field</li>
    <li><code class="language-plaintext highlighter-rouge">$2</code> stores second field and so on.</li>
    <li><code class="language-plaintext highlighter-rouge">NF</code> stores the total number of fields.</li>
  </ul>
</blockquote>

<p>Similarly, if we assign newline character as the field separator, 
then each line for that FASTA record will act as different columns.
The first column would be the description line, and the remaining columns
would be the sequences for that FASTA record.</p>

<p>This can be accomplished with option <code class="language-plaintext highlighter-rouge">-v</code> which allows setting a variable.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span> <span class="s1">'
    {$1=$1; print $0}
  '</span> files/fasta/example.fasta
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
Xanthomonas perforans strain GV872 contig_50    TGCGGGCGGTGAGGTCGCCCGCCGCGATCGCCTGCAATACCGTCGAGAGCGCGTGCAGATTCGCATCGGAGGTGGCCATC        AATTGATTGAGGCTGTCCACCATCGCGCGGAAGTCGTACTGGAAGCGCTGTGCGTCGCCGCGCACGCTGAAATCGCCATT    GGCTGCGGCCTGCGCAAGCAGCTTGATCTCGCTGTTCATCGCCGACAGGTTG
Xanthomonas perforans strain GV872 contig_51    CCCTCCACAGGAAGCGGACCGGACAATGACGCTTGATTGCCGCGCAGGAAATGCCCCCGTTCCCGCAGGCATACACCCCA        TTTGCCGCGGACATACGCCGCACATTTCTCGGAGCCTGGACAAGTCGCCACAGCTTTGATGGAACGACGGAAAAAACGCC    CCGGAGTTGACCAGAACAACCGCCAGTACTGAAATCTGGACCTTAAGTTCCAGTTCCTGCAGCTCTGGACCTGGTTCATC        AGCGAGCTTCTATCGCTTCGCGGCGGCTTCGATCGCAGCGTGTGCGCGCCATCACTTGGTCGTGCGAAATCGTCGGCCGC   GTGTCGGCCACGGCGCGCTCGATCTTCGCCCGGAACTACCAGTCGTAGTCCTCGGCGGCTTCGGTCGACTCAAATTCGGA AACAGGTGGTCGAGTGATACGCCCAGGGTTC
Xanthomonas perforans strain GV872 contig_52    CGCCAGCGCAGCCGACCACGACGCCCGAATCCAGCGTGCCGGCCTCGTTGCCGCCCGCGTCCACGTCGGGCAATTAAGCC        GCGCCCCGCGTTTCCGGTCATGGAACTTCGCCCAGAGCAGCTGGCCGGCCAGTCCAGCCAGCCATTGCAGCCGGTGTATC    TGATCGCCGGCCCCGAGACCTTGCGTGTGCTTGAGGCCGCCGATGCGGTGCGCGCGCGCGCGCGCGC
Xanthomonas perforans strain GV872 plasmid_contig_53    GGCGTTGTTCCAGCCCAGCACCATTTCCAACAGATTCAGGGCCTGCGCCGGCGACACCGGATTGGGCGTGGCCAAGGTCA        CGGTGCCCTGCACGCCCGGCGCGATGACGTAGTTCTGGCCGAGCATGTCGCCCAGGATCGCCTTGACCACCGCCTGCAGC    GATTCGCCTTCGAAATTGAAGGTGGCGCTGCCGCTGCTGGCCATGCCCAGCGTG
Xanthomonas perforans strain GV872 plasmid_contig_54    AAATCCTGGGCGTATCAGACCGCTGCCGAGCGTTTGCTTTGGTACCGCCTGCGTGATCGGCGTCCCCCC/àùbä';ç3cC
¦Å¼OÈ’LäÖ’sëmôzÒÒ${–-ÃêFY;+ã}:ãž0}%ÂZDÅì9D‹Rž÷D×ç°:âñÆó²9h‰ÐžæK¼ë™XÕJK¥·—(ü7¢]ÃÆ    Ño5®o&amp;ÇDçJ^ÃŒZòsÀš‹Ü´†ü¿ÄŒ—xÎ—¸¬µ»?à=Of±)2?îP3ñLapÈëY_{DŸm)ã}ž[Êº„ódMÅbóSl¤ÌŽÑ‚¼    Óh+¶\Éz½/­Z¹â²kgfSû%äÖÿzvŽ#?69éhÎÔøK×‰Êí¨/4ŠSVfÍõÔûÁÒ&amp;½ûE÷kÑ²:k ÷yxÙgH2/ÂžQ,[@ûIß    €|Ë²Ï2,ôCÆ
</code></pre></div></div>

<blockquote class="notes">
  <p><code class="language-plaintext highlighter-rouge">$1=$1</code> is used to recompute the records, so <code class="language-plaintext highlighter-rouge">OFS</code> can be applied.</p>
</blockquote>

<p>Notice the empty line in the first line of output. 
It exists because <code class="language-plaintext highlighter-rouge">awk</code> thinks there is a record before the first <code class="language-plaintext highlighter-rouge">&gt;</code>,
as it is a record separator, not record starter.
To avoid this empty line, we can ask <code class="language-plaintext highlighter-rouge">awk</code> to not process the first record.
This can be done with a another <code class="language-plaintext highlighter-rouge">awk</code> variable, <code class="language-plaintext highlighter-rouge">NR</code>. 
<code class="language-plaintext highlighter-rouge">NR</code> stands for number of records, and is roughly equivalent to 
row number or line number being processed.</p>

<p>Conditional execution of code can be done my writing the condition 
before the braces.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span> <span class="s1">'
    NR&gt;1 {$1=$1; print $0} # Only execute if record number is greater than 1.
  '</span> files/fasta/example.fasta

<span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span> <span class="s1">'
    NR==1 {next} # Skip to next record if record number is equal to 1.
    {$1=$1; print $0}
  '</span> files/fasta/example.fasta
</code></pre></div></div>

<h2 id="fasta-filtration">FASTA filtration</h2>

<p>If is often necessary to remove one or more FASTA record from a multiFASTA file.
Small files can be edited manually,
but large files or files with a large number of records
can be better handled programmatically.</p>

<p>FASTA filtration is often done using two criteria.</p>
<ol>
  <li>Sequence ID, i.e., the description line</li>
  <li>Length of the sequence</li>
</ol>

<h3 id="fasta-filtration-by-record-id">FASTA filtration by record ID</h3>

<p>FASTA filtration by ID is necessary in various cases such as removal of 
plasmid, plastid, or mitochondrial sequence from chromosomal sequence, 
removal of aberrant sequences before alignment etc.</p>

<p>In the code above, the sequence IDs can be accessed within <code class="language-plaintext highlighter-rouge">awk</code> with 
variable <code class="language-plaintext highlighter-rouge">$1</code>. Here are a few conditions for string matching.</p>
<ul>
  <li><code class="language-plaintext highlighter-rouge">$1=="keyword"</code>: sequence ID exactly matches <code class="language-plaintext highlighter-rouge">keyword</code> (except <code class="language-plaintext highlighter-rouge">&gt;</code> symbol).</li>
  <li><code class="language-plaintext highlighter-rouge">$1~/keyword/</code>: sequence ID contains <code class="language-plaintext highlighter-rouge">keyword</code>.</li>
  <li><code class="language-plaintext highlighter-rouge">$1~/^keyword/</code>: sequence ID starts with <code class="language-plaintext highlighter-rouge">keyword</code>.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">^</code> is resolved to start anchor.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">$1~/keyword$/</code>: sequence ID ends with <code class="language-plaintext highlighter-rouge">keyword</code>.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">$</code> is resolved to end anchor.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">$1!~/keyword/</code>: sequence ID does not contain <code class="language-plaintext highlighter-rouge">keyword</code>.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">!</code> is resolved to <strong>NOT</strong>.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">^.a</code>: sequence ID with <code class="language-plaintext highlighter-rouge">a</code> as the second character.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">.</code> can be resolved into any single character.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">^\.a</code>: sequence ID starts with <code class="language-plaintext highlighter-rouge">.a</code>.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">\</code> is an escape character, thus <code class="language-plaintext highlighter-rouge">.</code> is used in literal fashion.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">$1~/[a-c]/</code>: sequence ID contains <code class="language-plaintext highlighter-rouge">a</code>, <code class="language-plaintext highlighter-rouge">b</code>, or <code class="language-plaintext highlighter-rouge">c</code>.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">[...]</code> is resolved to character group.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">$1~/^[2-4]/</code>: sequence ID starts with <code class="language-plaintext highlighter-rouge">2</code>, <code class="language-plaintext highlighter-rouge">3</code>, or <code class="language-plaintext highlighter-rouge">4</code>.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">-</code> is resolved to character range.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">$1~/2[0-9]$/</code>: sequence ID ends between <code class="language-plaintext highlighter-rouge">20</code> and <code class="language-plaintext highlighter-rouge">29</code>.</li>
  <li><code class="language-plaintext highlighter-rouge">$1~/19$/ || $1~/2[0-1]$/</code>: sequence ID ends with <code class="language-plaintext highlighter-rouge">19</code>, <code class="language-plaintext highlighter-rouge">20</code>, or <code class="language-plaintext highlighter-rouge">21</code>.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">||</code> is resolved to <strong>OR</strong>.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">$1~/key1/ &amp;&amp; $1~/key2/</code>: sequence ID contains both <code class="language-plaintext highlighter-rouge">key1</code> and <code class="language-plaintext highlighter-rouge">key2</code> in no order.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">&amp;&amp;</code> is resolved to <strong>AND</strong>.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">$1~/key1.*key2/</code>: sequence ID contains both <code class="language-plaintext highlighter-rouge">key1</code> and <code class="language-plaintext highlighter-rouge">key2</code> with <code class="language-plaintext highlighter-rouge">key1</code> before <code class="language-plaintext highlighter-rouge">key2</code>.
    <ul>
      <li><code class="language-plaintext highlighter-rouge">.*</code> is resolved to any characters, including nothing.</li>
    </ul>
  </li>
  <li><code class="language-plaintext highlighter-rouge">$1~/^key1.*key2$/</code>: sequence ID starts with <code class="language-plaintext highlighter-rouge">key1</code> and ends with <code class="language-plaintext highlighter-rouge">key2</code>.</li>
</ul>

<p>The example FASTA above has two plasmid contigs, 
which can be separated with the term “plasmid” in description line.
To remove plasmid sequences, we can do:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">""</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="s1">'
    NR&gt;1 &amp;&amp; $1!~/plasmid/ {print "&gt;"$0}
  '</span> files/fasta/example.fasta
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;Xanthomonas perforans strain GV872 contig_50
TGCGGGCGGTGAGGTCGCCCGCCGCGATCGCCTGCAATACCGTCGAGAGCGCGTGCAGATTCGCATCGGAGGTGGCCATC
AATTGATTGAGGCTGTCCACCATCGCGCGGAAGTCGTACTGGAAGCGCTGTGCGTCGCCGCGCACGCTGAAATCGCCATT
GGCTGCGGCCTGCGCAAGCAGCTTGATCTCGCTGTTCATCGCCGACAGGTTG
&gt;Xanthomonas perforans strain GV872 contig_51
CCCTCCACAGGAAGCGGACCGGACAATGACGCTTGATTGCCGCGCAGGAAATGCCCCCGTTCCCGCAGGCATACACCCCA
TTTGCCGCGGACATACGCCGCACATTTCTCGGAGCCTGGACAAGTCGCCACAGCTTTGATGGAACGACGGAAAAAACGCC
CCGGAGTTGACCAGAACAACCGCCAGTACTGAAATCTGGACCTTAAGTTCCAGTTCCTGCAGCTCTGGACCTGGTTCATC
AGCGAGCTTCTATCGCTTCGCGGCGGCTTCGATCGCAGCGTGTGCGCGCCATCACTTGGTCGTGCGAAATCGTCGGCCGC
GTGTCGGCCACGGCGCGCTCGATCTTCGCCCGGAACTACCAGTCGTAGTCCTCGGCGGCTTCGGTCGACTCAAATTCGGA
AACAGGTGGTCGAGTGATACGCCCAGGGTTC
&gt;Xanthomonas perforans strain GV872 contig_52
CGCCAGCGCAGCCGACCACGACGCCCGAATCCAGCGTGCCGGCCTCGTTGCCGCCCGCGTCCACGTCGGGCAATTAAGCC
GCGCCCCGCGTTTCCGGTCATGGAACTTCGCCCAGAGCAGCTGGCCGGCCAGTCCAGCCAGCCATTGCAGCCGGTGTATC
TGATCGCCGGCCCCGAGACCTTGCGTGTGCTTGAGGCCGCCGATGCGGTGCGCGCGCGCGCGCGCGC
</code></pre></div></div>

<blockquote class="notes">
  <p>A <code class="language-plaintext highlighter-rouge">&gt;</code> was added to print statement to recover the <code class="language-plaintext highlighter-rouge">&gt;</code> in the front of
description line.</p>
</blockquote>

<h3 id="fasta-filtration-by-length">FASTA filtration by length</h3>

<p>FASTA filtration by length is necessary cases such as removal of contigs
smaller than a threshold length.</p>

<p>The numbers of characters in a FASTA record can be derived by <code class="language-plaintext highlighter-rouge">awk</code> function, 
<code class="language-plaintext highlighter-rouge">length</code>. 
Of course, this way, all characters, 
including those not standard in FASTA are counted.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span> <span class="s1">'
    NR&gt;1 {a=$1; $1 = ""; print a, length($0)-NF+1}
  '</span> files/fasta/example.fasta
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Xanthomonas perforans strain GV872 contig_50    212
Xanthomonas perforans strain GV872 contig_51    431
Xanthomonas perforans strain GV872 contig_52    227
Xanthomonas perforans strain GV872 plasmid_contig_53    214
Xanthomonas perforans strain GV872 plasmid_contig_54    331
</code></pre></div></div>

<blockquote class="notes">
  <h2 id="how-it-works">How it works</h2>
  <p><code class="language-plaintext highlighter-rouge">a=$1</code> stores description line in variable <code class="language-plaintext highlighter-rouge">a</code>.
<code class="language-plaintext highlighter-rouge">$1=""</code> empties description line, so <code class="language-plaintext highlighter-rouge">$0</code> only holds sequence only.
<code class="language-plaintext highlighter-rouge">length($0)</code> calculates length of the sequence.
<code class="language-plaintext highlighter-rouge">-NF+1</code> adjusts character length to string as <code class="language-plaintext highlighter-rouge">awk</code> also counts field separators.</p>
</blockquote>

<p>Next, we can do filtration.
To remove contigs smaller than 300 base pairs, we can do:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">""</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="s1">'
    NR&gt;1 {a=$0; $1 = ""; if(length-NF+1&gt;=300) {print "&gt;"a}}
  '</span> files/fasta/example.fasta
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;Xanthomonas perforans strain GV872 contig_51
CCCTCCACAGGAAGCGGACCGGACAATGACGCTTGATTGCCGCGCAGGAAATGCCCCCGTTCCCGCAGGCATACACCCCA
TTTGCCGCGGACATACGCCGCACATTTCTCGGAGCCTGGACAAGTCGCCACAGCTTTGATGGAACGACGGAAAAAACGCC
CCGGAGTTGACCAGAACAACCGCCAGTACTGAAATCTGGACCTTAAGTTCCAGTTCCTGCAGCTCTGGACCTGGTTCATC
AGCGAGCTTCTATCGCTTCGCGGCGGCTTCGATCGCAGCGTGTGCGCGCCATCACTTGGTCGTGCGAAATCGTCGGCCGC
GTGTCGGCCACGGCGCGCTCGATCTTCGCCCGGAACTACCAGTCGTAGTCCTCGGCGGCTTCGGTCGACTCAAATTCGGA
AACAGGTGGTCGAGTGATACGCCCAGGGTTC
&gt;Xanthomonas perforans strain GV872 plasmid_contig_54
AAATCCTGGGCGTATCAGACCGCTGCCGAGCGTTTGCTTTGGTACCGCCTGCGTGATCGGCGTCCCCCC/àùbä';ç3cC
¦Å¼OÈ’LäÖ’sëmôzÒÒ${–-ÃêFY;+ã}:ãž0}%ÂZDÅì9D‹Rž÷D×ç°:âñÆó²9h‰ÐžæK¼ë™XÕJK¥·—(ü7¢]ÃÆ
Ño5®o&amp;ÇDçJ^ÃŒZòsÀš‹Ü´†ü¿ÄŒ—xÎ—¸¬µ»?à=Of±)2?îP3ñLapÈëY_{DŸm)ã}ž[Êº„ódMÅbóSl¤ÌŽÑ‚¼
Óh+¶\Éz½/­Z¹â²kgfSû%äÖÿzvŽ#?69éhÎÔøK×‰Êí¨/4ŠSVfÍõÔûÁÒ&amp;½ûE÷kÑ²:k ÷yxÙgH2/ÂžQ,[@ûIß
€|Ë²Ï2,ôCÆ
</code></pre></div></div>

<blockquote class="notes">
  <p><code class="language-plaintext highlighter-rouge">if</code> is used for writing a conditional statement.</p>
</blockquote>

<p>If the FASTA contains non-standard characters, and we only want to count and 
filter using standard bases, we can use <code class="language-plaintext highlighter-rouge">gsub</code> function.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span> <span class="s1">'
    NR&gt;1 {a=$1; $1 = ""; print a, gsub(/[aAtTgGcCnN]/, $0)}
  '</span> files/fasta/example.fasta

<span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">""</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="s1">'
    NR&gt;1 {a=$0; $1 = ""; if(gsub(/[aAtTgGcCnN]/, $0)&gt;=300) {print "&gt;"a}}
  '</span> files/fasta/example.fasta
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Xanthomonas perforans strain GV872 contig_50    212
Xanthomonas perforans strain GV872 contig_51    431
Xanthomonas perforans strain GV872 contig_52    227
Xanthomonas perforans strain GV872 plasmid_contig_53    214
Xanthomonas perforans strain GV872 plasmid_contig_54    75

&gt;Xanthomonas perforans strain GV872 contig_51
CCCTCCACAGGAAGCGGACCGGACAATGACGCTTGATTGCCGCGCAGGAAATGCCCCCGTTCCCGCAGGCATACACCCCA
TTTGCCGCGGACATACGCCGCACATTTCTCGGAGCCTGGACAAGTCGCCACAGCTTTGATGGAACGACGGAAAAAACGCC
CCGGAGTTGACCAGAACAACCGCCAGTACTGAAATCTGGACCTTAAGTTCCAGTTCCTGCAGCTCTGGACCTGGTTCATC
AGCGAGCTTCTATCGCTTCGCGGCGGCTTCGATCGCAGCGTGTGCGCGCCATCACTTGGTCGTGCGAAATCGTCGGCCGC
GTGTCGGCCACGGCGCGCTCGATCTTCGCCCGGAACTACCAGTCGTAGTCCTCGGCGGCTTCGGTCGACTCAAATTCGGA
AACAGGTGGTCGAGTGATACGCCCAGGGTTC
</code></pre></div></div>

<blockquote class="tips">
  <p>Find the list of all IUPAC nucleotide and amino acid codes <a href="https://en.wikipedia.org/wiki/International_Union_of_Pure_and_Applied_Chemistry#Amino_acid_and_nucleotide_base_codes">here</a>.</p>
</blockquote>

<blockquote class="challenge">
  <h2 id="exercise-title">Exercise title</h2>

  <p>Write a <code class="language-plaintext highlighter-rouge">awk</code> code to extract first 2 FASTA records from a multiFASTA file and
write it to file</p>

  <details>
    <p><!--collapsible region--></p>
    <summary></summary>

    <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># 1</span>
<span class="gp">$ </span><span class="nb">awk</span> <span class="nt">-vRS</span><span class="o">=</span><span class="s2">"&gt;"</span> <span class="nt">-vORS</span><span class="o">=</span><span class="s2">""</span> <span class="nt">-vFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="nt">-vOFS</span><span class="o">=</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> <span class="s1">'
   NR&gt;1 &amp;&amp; NR &lt;=3 {$1=$1; print "&gt;"$0}
 '</span> files/fasta/example.fasta
</code></pre></div>    </div>

  </details>
</blockquote>

<p>Using what you learned from optional lessons 1 and 2, 
you can now write a script to filter FASTA files and distribute it to others.</p>

</article>



























  
  
















<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../10-more_unix/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../30-extra_tools/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>


      
      
<footer>
  <hr/>
  <div class="row">
    <div class="col-md-6" align="left">
      <h4>
	      Copyright &copy; 2020–2022 <a href="https://github.com/BioInfoAPS">APS Basic Bioinformatics Workshop Team</a>
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>

<script>
/* Based on https://stackoverflow.com/a/30894829/1802726 */
var show_instructions = function() {
     //set all nav tabs to inactive
    $('.nav-tabs li').removeClass('active');

    //get all nav tabs matching the os and set to active
    $('.nav-tabs li a[href*="'+localStorage.os+'"]').closest('li').addClass('active');

    //set all tabs to inactive
    $('.tab-pane').removeClass('active');

    //get all tabs matching the os and set to active
    $('.tab-pane[id*="'+localStorage.os+'"]').addClass('active');
}

if (typeof(Storage) !== "undefined") {
    if (localStorage.os == "undefined") {
        localStorage.os = "windows";
    }
    show_instructions();
} else {
    console.log("No localStorage");
}

/* Based on https://stackoverflow.com/a/30894829/1802726 */
$('.nav-tabs li a').click(function(){
    //get selected os
    localStorage.os = $(this)[0].dataset.os;
    show_instructions();
});

  // detect misssing iframe for Eventbrite
  $(document).ready(function(){
      if ($('iframe').height() == 0) {
          $('iframe').before('<p>Looks like your adblocker blocked the registration window. Please navigate to <a href="https://www.eventbrite.com/tickets-external?eid=&ref=etckt">https://www.eventbrite.com/tickets-external?eid=&ref=etckt</a> to register.</p>');
      }
  });

  // hide alerts when site is rendered on carpentries.github.io
  $(document).ready(function() {
      if (location.href.startsWith("https://carpentries.github.io/workshop-template/")) {
          $("div.alert").hide();
      }
  });

</script>

  </body>
</html>
